<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="config_rules_RIL.yaml" href="config_rules_RIL.html" /><link rel="prev" title="Test_yaml_detectCOs_BC_EMS.yaml" href="Test_yaml_detectCOs_BC_EMS.html" />

    <!-- Generated with Sphinx 5.3.0 and Furo 2023.03.27 -->
        <title>detectCOs_identifyCOs_RIL.py - detectCOs 2.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=fad236701ea90a88636c2a8c73b44ae642ed2a53" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">detectCOs 2.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">detectCOs 2.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="config_detectCOs_CG.html">config_detectCOs_CG.yaml</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_required_functions.html">detectCOs_required_functions.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_read_files.html">detectCOs_read_files.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_sliding_window.html">detectCOs_sliding_window.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_identifyCOs.html">detectCOs_identifyCOs.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="test_detectCOs_polyrec_functions.html">test_detectCOs_polyrec_functions.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="Test_laucher_With_diff_SNPs.html">Test_laucher_With_diff_SNPs.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="VisualiseGenotypesChromosome.html">VisualiseGenotypesChromosome.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_identifyCOs_EMS.html">detectCOs_identifyCOs_EMS.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_required_functions_EMS.html">detectCOs_required_functions_EMS.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_sliding_window_EMS.html">detectCOs_sliding_window_EMS.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="Test_yaml_detectCOs_BC_EMS.html">Test_yaml_detectCOs_BC_EMS.yaml</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">detectCOs_identifyCOs_RIL.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="config_rules_RIL.html">config_rules_RIL.yaml</a></li>
<li class="toctree-l1"><a class="reference internal" href="Launcher_Rules_RIL.html">Launcher_Rules_RIL.py</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="detectcos-identifycos-ril-py">
<h1>detectCOs_identifyCOs_RIL.py<a class="headerlink" href="#detectcos-identifycos-ril-py" title="Permalink to this heading">#</a></h1>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">hues</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<span class="kn">from</span> <span class="nn">detectCOs_required_functions</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
<section id="function-identifycos-rils">
<h2>Function : IdentifyCOs_RILs<a class="headerlink" href="#function-identifycos-rils" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">logging</span>

<span class="n">logging</span><span class="o">.</span><span class="n">basicConfig</span><span class="p">(</span><span class="n">level</span><span class="o">=</span><span class="n">logging</span><span class="o">.</span><span class="n">INFO</span><span class="p">,</span> <span class="nb">format</span><span class="o">=</span><span class="s1">&#39;</span><span class="si">%(message)s</span><span class="s1">&#39;</span><span class="p">)</span>
<span class="n">logger</span> <span class="o">=</span> <span class="n">logging</span><span class="o">.</span><span class="n">getLogger</span><span class="p">()</span>

<span class="k">def</span> <span class="nf">log</span><span class="p">(</span><span class="n">message</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&#39;&#39;&#39;</span>
<span class="sd">    Écrit un message à la fois dans le fichier de log et sur la console.</span>
<span class="sd">    &#39;&#39;&#39;</span>
    <span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="n">message</span><span class="p">)</span>

<span class="k">def</span> <span class="nf">IdentifyCOs_RILs</span><span class="p">(</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">,</span> <span class="n">nb_windows_chr</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:    Find location of hypothetical cross-over according to the change of genotype</span>

<span class="sd">    Input:                  - offspring_genotype_window_smoothed: Ordered dictionnary</span>
<span class="sd">                    from SmoothNormalizedOsffspringSlidingWindow containing :</span>
<span class="sd">                    [chr_window] = [start, stop, ADref/DP, ADalt/DP, probHomoRef, probHetero, probHomoAlt, genotype]</span>

<span class="sd">                    - nb_windows_chr:       number of window per chromosome, determine with function</span>
<span class="sd">                    NormalizeOffspringSlidingWindow()</span>

<span class="sd">    Output:                 Cross-over candidate from smoothed windows</span>
<span class="sd">                    candidates_co[chr_mean_window] = [left_window_border + &quot;:&quot; + right_window_border,</span>
<span class="sd">                                                    co pos start, co pos stop, previous genotype, new genotype]</span>

<span class="sd">                    All double cross-over found during the analysis are saved, even if they are under the threshold</span>
<span class="sd">                    to support the genotype (thr_support) :</span>
<span class="sd">                    db_co[chr_mean_window] = [cur_chr, start_co_win:stop_co_win, previous genotype, new genotype, status]</span>

<span class="sd">    Upgrade? :      add input parameter : length required to support genotype (kb, default 300kb) and window size (kb)</span>
<span class="sd">                modify :</span>
<span class="sd">                    if stop_win &lt;= 10 or start_win &gt; (nb_win - 10):</span>
<span class="sd">                        thr_support = int((thr_support_kb/window_size) - 1) # (thr_support_kb/2)/(window_size/2) - 1</span>
<span class="sd">                    else:</span>
<span class="sd">                        thr_support = int((thr_support_kb/(window_size/2)) - 1)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">print</span><span class="p">()</span>
    <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Identify candidate crossovers&quot;</span><span class="p">)</span>
    <span class="n">log</span><span class="p">(</span><span class="s2">&quot;Identify candidate crossovers&quot;</span><span class="p">)</span>

    <span class="n">candidates_co</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">db_co</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span> <span class="c1"># save all homo to another homo genotype even if they are not enough supported</span>
    <span class="n">thr_support</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># number of windows to support genotype = threshold support</span>

    <span class="k">for</span> <span class="n">cur_chr</span><span class="p">,</span> <span class="n">nb_win</span> <span class="ow">in</span> <span class="n">nb_windows_chr</span><span class="o">.</span><span class="n">items</span><span class="p">():</span> <span class="c1">## iteration on each chromosome</span>
        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Processing chromosome: </span><span class="si">{</span><span class="n">cur_chr</span><span class="si">}</span><span class="s2"> with </span><span class="si">{</span><span class="n">nb_win</span><span class="si">}</span><span class="s2"> windows.&quot;</span><span class="p">)</span>

        <span class="c1"># information about the window num_win</span>
        <span class="n">cur_support</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># number of window that support current genotype</span>
        <span class="n">pre_num</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># number of window that support current genotype after a NA transition # modif MY</span>
        <span class="n">cur_geno</span> <span class="o">=</span> <span class="nb">str</span><span class="p">()</span> <span class="c1"># genotype of current window(=geno_stop)</span>
        <span class="n">cur_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># current window = num_win + 1</span>

        <span class="c1"># information about the previous supported geonotype</span>
        <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># genotype of the window before the hypothetical COs</span>
        <span class="n">pre_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># last window of the previous genotype (before hypothetical COs)</span>

        <span class="c1"># count between previous genotype and current genotype</span>
        <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># number of consecutive NA window</span>
        <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># support of genotype before starting NA region</span>
        <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># initialize geno_preNA at the start of processing each chromosome # modify MY</span>
        <span class="n">win_preNA</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># test MY</span>


        <span class="n">chrom</span> <span class="o">=</span> <span class="p">[]</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">offspring_genotype_window_smoothed</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="k">if</span> <span class="n">cur_chr</span> <span class="o">==</span> <span class="n">key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="n">chrom</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s1">&#39;_&#39;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>

        <span class="c1">#nb_win = int(len(chrom))</span>
        <span class="n">nb_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">chrom</span><span class="p">[</span><span class="o">-</span><span class="mi">1</span><span class="p">])</span>
        <span class="c1">#print(nb_win)</span>


        <span class="k">for</span> <span class="n">num_win</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="nb">len</span><span class="p">(</span><span class="n">chrom</span><span class="p">)</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span> <span class="c1"># iteration on the each window of each chromosome cur_chr</span>

            <span class="c1">#print(num_win)</span>


            <span class="c1"># Define genotype for window start and stop</span>
            <span class="n">start_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">chrom</span><span class="p">[</span><span class="n">num_win</span><span class="p">])</span> <span class="c1"># from 1 to the penultimate num_win (due to range function)</span>
            <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;</span><span class="se">\n</span><span class="s2">Analyzing window </span><span class="si">{</span><span class="n">start_win</span><span class="si">}</span><span class="s2">/</span><span class="si">{</span><span class="n">nb_win</span><span class="si">}</span><span class="s2"> on chromosome </span><span class="si">{</span><span class="n">cur_chr</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
            <span class="n">key_start</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">start_win</span><span class="p">)</span>
            <span class="n">geno_start</span> <span class="o">=</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">key_start</span><span class="p">][</span><span class="mi">9</span><span class="p">]</span>

            <span class="n">stop_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">chrom</span><span class="p">[</span><span class="n">num_win</span> <span class="o">+</span> <span class="mi">1</span><span class="p">])</span><span class="c1"># from 2 to the last num_win</span>
            <span class="n">key_stop</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_win</span><span class="p">)</span>
            <span class="n">geno_stop</span> <span class="o">=</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">key_stop</span><span class="p">][</span><span class="mi">9</span><span class="p">]</span>

            <span class="c1">## Edit the thr_support = min number of window to validate a COs   # modify MY</span>
            <span class="k">if</span> <span class="n">start_win</span> <span class="o">&lt;=</span> <span class="mi">10</span><span class="p">:</span>
                <span class="c1"># for 9 first  slides   # modify MY</span>
                <span class="k">if</span> <span class="n">stop_win</span> <span class="o">==</span> <span class="mi">11</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>   <span class="c1"># modify MY</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">5</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="k">elif</span> <span class="n">start_win</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">10</span><span class="p">):</span>
                <span class="c1"># for 9 last slides</span>
                <span class="k">if</span> <span class="n">start_win</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">10</span><span class="p">)</span> <span class="ow">and</span> <span class="n">start_win</span> <span class="o">&lt;=</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">6</span><span class="p">)</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">10</span><span class="p">)][</span><span class="mi">9</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>  <span class="c1"># modify MY</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">5</span>
                <span class="k">elif</span> <span class="n">start_win</span> <span class="o">==</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">10</span><span class="p">)</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">!=</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">5</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="k">else</span><span class="p">:</span> <span class="c1"># for all other windows</span>
                <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">5</span>

            <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Window start genotype: </span><span class="si">{</span><span class="n">geno_start</span><span class="si">}</span><span class="s2">, Window stop genotype: </span><span class="si">{</span><span class="n">geno_stop</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>

            <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Current threshold support: </span><span class="si">{</span><span class="n">thr_support</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
            <span class="c1">## Initialization of the start and stop posistion of the CO</span>
            <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># mean position of the start window of the crossover</span>
            <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># mean position of the stop window of the crossover</span>
            <span class="n">stop_co_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># number of window of the end of the crossover region</span>


            <span class="k">if</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="n">geno_stop</span> <span class="p">:</span>
                <span class="n">log</span><span class="p">(</span><span class="s2">&quot;Genotype change detected between consecutive windows.&quot;</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">or</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="n">log</span><span class="p">(</span><span class="s2">&quot;Handling NA genotype scenario...&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="p">:</span>
                        <span class="n">log</span><span class="p">(</span><span class="s2">&quot;Handling geno_stop == &#39;NA&#39; and  geno_start != &#39;NA&#39; scenario...&quot;</span><span class="p">)</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;geno_start : </span><span class="si">{</span><span class="n">geno_start</span><span class="si">}</span><span class="s1">  , geno_stop :  </span><span class="si">{</span><span class="n">geno_stop</span><span class="si">}</span><span class="s1"> &#39;</span><span class="p">)</span>
                        <span class="c1"># when enter in NA region:</span>
                        <span class="n">geno_preNA</span> <span class="o">=</span> <span class="n">geno_start</span>
                        <span class="n">win_preNA</span> <span class="o">=</span> <span class="n">start_win</span>
                        <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="n">support_preNA</span> <span class="o">=</span> <span class="n">cur_support</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;geno_preNA = </span><span class="si">{</span><span class="n">geno_preNA</span><span class="si">}</span><span class="s1">, win_preNA = </span><span class="si">{</span><span class="n">win_preNA</span><span class="si">}</span><span class="s1">, count_NA = </span><span class="si">{</span><span class="n">count_NA</span><span class="si">}</span><span class="s1">, support_preNA = </span><span class="si">{</span><span class="n">support_preNA</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                        <span class="k">continue</span>

                    <span class="k">elif</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="n">geno_preNA</span> <span class="p">:</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Handling geno_start == &#39;NA&#39; and  geno_stop (</span><span class="si">{</span><span class="n">geno_stop</span><span class="si">}</span><span class="s2">) == geno_preNA (</span><span class="si">{</span><span class="n">geno_preNA</span><span class="si">}</span><span class="s2">) scenario...&quot;</span><span class="p">)</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;pre_win = </span><span class="si">{</span><span class="n">pre_win</span><span class="si">}</span><span class="s1">, pre_geno = </span><span class="si">{</span><span class="n">pre_geno</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                        <span class="n">cur_support</span> <span class="o">+=</span><span class="mi">1</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;cur_support = </span><span class="si">{</span><span class="n">cur_support</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                        <span class="n">support_preNA</span> <span class="o">=</span> <span class="n">cur_support</span>  <span class="c1"># modify MY</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;support_preNA = </span><span class="si">{</span><span class="n">support_preNA</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>


                        <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&gt;=</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">!=</span>  <span class="n">geno_stop</span><span class="p">:</span>      <span class="c1">## voir cas 13   # modif MY</span>
                            <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>
                            <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
                            <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>

                            <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                            <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                            <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                            <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                            <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                            <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;key </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> supported by </span><span class="si">{</span><span class="n">cur_support</span><span class="si">}</span><span class="s2"> in candidateCO with NA.&quot;</span><span class="p">)</span>
                            <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

                            <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                                <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>

                            <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>                             <span class="c1"># modify MY</span>
                            <span class="n">pre_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span>                           <span class="c1"># test MY</span>
                            <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>                              <span class="c1"># modify MY</span>
                            <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span>                         <span class="c1"># modify MY</span>
                            <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>  <span class="c1"># Also reset geno_preNA  # modify MY</span>

                        <span class="k">continue</span>

                    <span class="k">elif</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">!=</span> <span class="n">geno_preNA</span><span class="p">:</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Handling geno_start == &#39;NA&#39; and  geno_stop (</span><span class="si">{</span><span class="n">geno_stop</span><span class="si">}</span><span class="s2">) != geno_preNA (</span><span class="si">{</span><span class="n">geno_preNA</span><span class="si">}</span><span class="s2">) scenario...&quot;</span><span class="p">)</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;pre_win = </span><span class="si">{</span><span class="n">pre_win</span><span class="si">}</span><span class="s1">, pre_geno = </span><span class="si">{</span><span class="n">pre_geno</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;cur_support = </span><span class="si">{</span><span class="n">cur_support</span><span class="si">}</span><span class="s1">, thr_support = </span><span class="si">{</span><span class="n">thr_support</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                        <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;win_preNA = </span><span class="si">{</span><span class="n">win_preNA</span><span class="si">}</span><span class="s1">, count_NA = </span><span class="si">{</span><span class="n">count_NA</span><span class="si">}</span><span class="s1">, pre_num = </span><span class="si">{</span><span class="n">pre_num</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                        <span class="k">if</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>
                            <span class="c1"># if geno_stop != geno_preNA, edit pre_geno and pre_win with the information before NA region</span>
                            <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">geno_preNA</span> <span class="c1"># genotype before NA region</span>
                            <span class="n">pre_win</span> <span class="o">=</span> <span class="n">win_preNA</span> <span class="c1"># last window before NA region</span>
                            <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                            <span class="n">win_preNA</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39; test test&#39;</span><span class="p">)</span> <span class="c1"># test MY</span>
                            <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;cur_geno : </span><span class="si">{</span><span class="n">cur_geno</span><span class="si">}</span><span class="s1">, cur_win : </span><span class="si">{</span><span class="n">cur_win</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span> <span class="c1"># test MY</span>
                            <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">pre_geno</span>  <span class="c1"># test MY</span>
                            <span class="n">pre_win</span> <span class="o">=</span> <span class="n">pre_win</span> <span class="c1"># test MY</span>
                            <span class="n">cur_support</span> <span class="o">=</span> <span class="n">pre_num</span> <span class="c1"># test MY</span>
                            <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span> <span class="c1"># test MY</span>
                            <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span> <span class="c1"># test MY</span>
                            <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># test MY</span>
                            <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># test MY</span>
                            <span class="k">continue</span> <span class="c1"># test MY</span>






                <span class="k">elif</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>
                    <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cur_support = </span><span class="si">{</span><span class="n">cur_support</span><span class="si">}</span><span class="s2">, in the first side.&quot;</span><span class="p">)</span>
                    <span class="c1"># if genotype change between start and stop num_win and</span>
                    <span class="c1"># genotype supported by at least threshold support (2 or 5)</span>
                    <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">geno_start</span> <span class="c1"># genotype of window before CO</span>
                    <span class="n">pre_win</span> <span class="o">=</span> <span class="n">start_win</span> <span class="c1"># window before CO</span>
                    <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span>                                        <span class="c1"># modify MY</span>
                    <span class="n">pre_num</span> <span class="o">=</span> <span class="n">cur_support</span> <span class="c1"># useless  # modify MY</span>

                <span class="n">cur_support</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># reset cur_support</span>

                <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">geno_start</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">geno_stop</span><span class="p">)</span> <span class="ow">and</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="n">key_db_co</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">((</span><span class="n">start_win</span> <span class="o">+</span> <span class="n">stop_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                    <span class="n">db_co</span><span class="p">[</span><span class="n">key_db_co</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">start_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_win</span><span class="p">),</span> <span class="n">geno_start</span><span class="p">,</span> <span class="n">geno_stop</span><span class="p">,</span> <span class="s2">&quot;not_supported&quot;</span><span class="p">]</span>
                    <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;We are here </span><span class="si">{</span><span class="n">key_db_co</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
                    <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;pre_win = </span><span class="si">{</span><span class="n">pre_win</span><span class="si">}</span><span class="s1">, pre_geno = </span><span class="si">{</span><span class="n">pre_geno</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>
                    <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s1">&#39;geno_start = </span><span class="si">{</span><span class="n">geno_start</span><span class="si">}</span><span class="s1">= pre_geno, geno_stop = </span><span class="si">{</span><span class="n">geno_stop</span><span class="si">}</span><span class="s1">&#39;</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>  <span class="c1"># Added by MY</span>
                    <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
                    <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>
                    <span class="n">log</span><span class="p">(</span><span class="s2">&quot;Crossover candidate detected due to genotype change and validate thr_support in last window.&quot;</span><span class="p">)</span>
                    <span class="c1">#log(&quot;Crossover candidate detected due to genotype change and validate thr_support in last window.&quot;)</span>
                    <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&lt;</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">support_preNA</span> <span class="o">!=</span> <span class="mi">0</span> <span class="p">:</span>
                        <span class="c1"># if there is some NA window during the count of cur_support</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="k">else</span> <span class="p">:</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">+</span> <span class="mi">1</span>

                    <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                    <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                    <span class="c1">## Edit candidate</span>
                    <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                    <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;key </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> supported by </span><span class="si">{</span><span class="n">cur_support</span><span class="si">}</span><span class="s2"> in candidateCO at last window.&quot;</span><span class="p">)</span>
                    <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>
                    <span class="c1">#log(f&quot;key {key} supported by {cur_support} in candidateCO at last window.&quot;)</span>

                    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                        <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>

                    <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">pre_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span>                           <span class="c1"># test MY</span>

            <span class="k">else</span><span class="p">:</span> <span class="c1"># geno_start == geno_stop</span>
                <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
                <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>

                <span class="k">if</span> <span class="n">cur_geno</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="p">:</span>
                    <span class="n">count_NA</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">continue</span>

                <span class="n">cur_support</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;Genotype </span><span class="si">{</span><span class="n">geno_stop</span><span class="si">}</span><span class="s2"> supported by </span><span class="si">{</span><span class="n">cur_support</span><span class="si">}</span><span class="s2"> consecutive windows.&quot;</span><span class="p">)</span>
                <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;pre_geno </span><span class="si">{</span><span class="n">pre_geno</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
                <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;cur_geno </span><span class="si">{</span><span class="n">cur_geno</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>
                <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;thr_support </span><span class="si">{</span><span class="n">thr_support</span><span class="si">}</span><span class="s2">.&quot;</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">cur_geno</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="k">elif</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span> <span class="c1"># modify MY</span>
                    <span class="n">log</span><span class="p">(</span><span class="s2">&quot;Crossover candidate detected due to genotype change and validate thr_support.&quot;</span><span class="p">)</span>
                    <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&lt;</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">support_preNA</span> <span class="o">!=</span> <span class="mi">0</span> <span class="p">:</span>
                        <span class="c1"># if there is some NA window during the count of cur_support</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="k">else</span> <span class="p">:</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">+</span> <span class="mi">1</span>

                    <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                    <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                    <span class="c1">## Edit candidate</span>
                    <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                    <span class="n">log</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;key </span><span class="si">{</span><span class="n">key</span><span class="si">}</span><span class="s2"> supported by </span><span class="si">{</span><span class="n">cur_support</span><span class="si">}</span><span class="s2"> in candidateCO without NA.&quot;</span><span class="p">)</span>
                    <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

                    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                        <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>

                    <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1">### Raise warning for all double COs found in the data</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">db_co</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Be careful, some double COs has been found during the analysis :&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">db_co</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1"># value = [cur_chr, start_co_win:stop_co_win, previous genotype, new genotype, status]</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- &quot;</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="s2">&quot; (window &quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;): &quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> \
                <span class="s2">&quot; (&quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="s2">&quot;)&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="n">log</span><span class="p">(</span><span class="s2">&quot;- &quot;</span> <span class="o">+</span> <span class="n">key</span> <span class="o">+</span> <span class="s2">&quot; (window &quot;</span> <span class="o">+</span> <span class="n">value</span><span class="p">[</span><span class="mi">1</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;): &quot;</span> <span class="o">+</span> <span class="n">value</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;&gt;&quot;</span> <span class="o">+</span> <span class="n">value</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot; (&quot;</span> <span class="o">+</span> <span class="n">value</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span> <span class="o">+</span> <span class="s2">&quot;)&quot;</span><span class="p">)</span>

    <span class="k">return</span> <span class="n">candidates_co</span><span class="p">,</span> <span class="n">db_co</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>IdentifyCOs_RILs</cite>: This function searches for the locations of hypothetical crossover events based on changes in genotype.</p>
<ul class="simple">
<li><p><strong>Input:</strong>
- <cite>offspring_genotype_window_smoothed</cite>: An ordered dictionary produced by the <cite>SmoothNormalizedOffspringSlidingWindow</cite> function containing smoothed genotype data of the offspring.
- <cite>nb_windows_chr</cite>: The number of windows per chromosome, determined using the <cite>NormalizeOffspringSlidingWindow</cite> function.</p></li>
<li><p><strong>Output:</strong>
- <cite>candidates_co</cite>: Candidates for crossovers, derived from the smoothed windows.
- <cite>db_co</cite>: All double crossovers found during analysis, even if they are not sufficiently supported.</p></li>
<li><p><strong>Functionality:</strong>
- The function iterates through each chromosome and examines each genotype data window.
- It detects changes in genotype between windows and identifies the start and end positions of the crossovers.
- Crossovers are recorded with their average positions and the genotypes before and after the crossover.
- Double crossovers are also recorded, even if they do not have sufficient support.</p></li>
</ul>
<p>The function also supports parameters to adjust the threshold of support needed to validate a crossover, as well as the required size to support a genotype.</p>
</section>
<section id="function-rules-rils">
<h2>Function : Rules_RIls<a class="headerlink" href="#function-rules-rils" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">Rules_RIls</span><span class="p">(</span><span class="n">info_RIL</span><span class="p">,</span><span class="n">offspring_EMS</span><span class="p">,</span><span class="n">offspring_basic</span> <span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Analyzes genotype data across multiple datasets to determine genetic rules and discrepancies.</span>

<span class="sd">    This function cross-references genotype data from basic and EMS offspring datasets with RIL information,</span>
<span class="sd">    identifies overlapping windows, and evaluates genotype discrepancies to establish a final genotype</span>
<span class="sd">    decision with associated warnings based on predefined rules.</span>

<span class="sd">    Parameters:</span>
<span class="sd">        info_RIL (dict): Dictionary containing RIL genotype information with window identifiers as keys.</span>
<span class="sd">        offspring_EMS (dict): Dictionary with EMS window data, where keys are window identifiers and</span>
<span class="sd">                            values include start and stop positions and genotypes.</span>
<span class="sd">        offspring_basic (dict): Dictionary similar to offspring_EMS but for basic genotype data.</span>

<span class="sd">    Returns:</span>
<span class="sd">        tuple: A tuple containing two dictionaries:</span>
<span class="sd">            1. RULEs (OrderedDict): Keys are window identifiers from offspring_basic and values are</span>
<span class="sd">                lists containing window start, stop, additional SNP info, and final genotype decisions.</span>
<span class="sd">            2. nb_window_chr (OrderedDict): Records the highest window number for each chromosome.</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">RULEs</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">nb_window_chr</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="k">for</span> <span class="n">EMS_key</span><span class="p">,</span> <span class="n">EMS_value</span> <span class="ow">in</span> <span class="n">offspring_EMS</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">start_EMS</span><span class="p">,</span> <span class="n">stop_EMS</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span> <span class="o">=</span> <span class="n">EMS_value</span>
        <span class="n">genotype_EMS</span> <span class="o">=</span> <span class="n">EMS_value</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>

        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">snp_info</span> <span class="ow">in</span> <span class="n">offspring_basic</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">candidate</span> <span class="o">=</span> <span class="n">key</span>
            <span class="n">chromosome</span><span class="o">=</span> <span class="n">candidate</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
            <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="o">*</span><span class="n">_</span><span class="p">,</span> <span class="n">genotype</span> <span class="o">=</span> <span class="n">snp_info</span>
            <span class="k">if</span> <span class="n">chromosome</span> <span class="o">==</span> <span class="n">EMS_key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]:</span>
                <span class="k">if</span> <span class="n">start_EMS</span> <span class="o">&lt;=</span> <span class="n">stop</span> <span class="ow">and</span> <span class="n">start</span> <span class="o">&lt;=</span> <span class="n">stop_EMS</span> <span class="p">:</span>

                    <span class="k">if</span> <span class="n">candidate</span> <span class="ow">in</span> <span class="n">info_RIL</span> <span class="ow">and</span> <span class="n">info_RIL</span><span class="p">[</span><span class="n">candidate</span><span class="p">][:</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">]:</span>
                        <span class="n">genotype_RIL</span> <span class="o">=</span> <span class="n">info_RIL</span><span class="p">[</span><span class="n">candidate</span><span class="p">][</span><span class="mi">18</span><span class="p">]</span>

                    <span class="n">genotype_EMS</span> <span class="o">=</span> <span class="n">EMS_value</span><span class="p">[</span><span class="mi">4</span><span class="p">]</span>

                    <span class="n">response</span> <span class="o">=</span> <span class="s2">&quot;Yes&quot;</span> <span class="k">if</span> <span class="n">genotype_EMS</span> <span class="o">==</span> <span class="s2">&quot;A&quot;</span> <span class="k">else</span> <span class="s2">&quot;No&quot;</span>

                    <span class="k">if</span> <span class="n">genotype_RIL</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col/Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;GT_NA&quot;</span>
                            <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                    <span class="k">elif</span> <span class="n">genotype_RIL</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;NEG_A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>                      <span class="c1"># modif</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col/Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;FALSE_A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>   <span class="c1"># modif</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;GT_NA&quot;</span>
                            <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL_NA&quot;</span>
                        <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>

                    <span class="n">RULEs</span><span class="p">[</span><span class="n">candidate</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span><span class="n">stop</span><span class="p">,</span><span class="n">snp_info</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">snp_info</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">genotype</span><span class="p">,</span> <span class="n">genotype_EMS</span><span class="p">,</span> <span class="n">genotype_RIL</span><span class="p">,</span><span class="n">response</span><span class="p">,</span><span class="n">WARNING</span><span class="p">,</span><span class="n">GT_Final</span><span class="p">]</span>

                <span class="k">elif</span> <span class="p">(</span><span class="n">start</span> <span class="o">&lt;=</span> <span class="n">offspring_EMS</span><span class="p">[</span><span class="n">chromosome</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="ow">and</span> <span class="n">stop</span> <span class="o">&lt;=</span> <span class="n">offspring_EMS</span><span class="p">[</span><span class="n">chromosome</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]):</span>
                    <span class="k">if</span> <span class="n">candidate</span> <span class="ow">in</span> <span class="n">info_RIL</span> <span class="ow">and</span> <span class="n">info_RIL</span><span class="p">[</span><span class="n">candidate</span><span class="p">][:</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">]:</span>
                            <span class="n">genotype_RIL</span> <span class="o">=</span> <span class="n">info_RIL</span><span class="p">[</span><span class="n">candidate</span><span class="p">][</span><span class="mi">18</span><span class="p">]</span>
                    <span class="n">genotype_EMS</span> <span class="o">=</span> <span class="n">offspring_EMS</span><span class="p">[</span><span class="n">chromosome</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="mi">1</span><span class="p">)][</span><span class="mi">4</span><span class="p">]</span>
                    <span class="n">response</span> <span class="o">=</span> <span class="s2">&quot;Yes&quot;</span> <span class="k">if</span> <span class="n">genotype_EMS</span> <span class="o">==</span> <span class="s2">&quot;A&quot;</span> <span class="k">else</span> <span class="s2">&quot;No&quot;</span>

                    <span class="k">if</span> <span class="n">genotype_RIL</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col/Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;GT_NA&quot;</span>
                            <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                    <span class="k">elif</span> <span class="n">genotype_RIL</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;NEG_A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>                      <span class="c1"># modif</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col/Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;FALSE_A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>   <span class="c1"># modif</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;GT_NA&quot;</span>
                            <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL_NA&quot;</span>
                        <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>

                    <span class="n">RULEs</span><span class="p">[</span><span class="n">candidate</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span><span class="n">stop</span><span class="p">,</span><span class="n">snp_info</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">snp_info</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">genotype</span><span class="p">,</span> <span class="n">genotype_EMS</span><span class="p">,</span> <span class="n">genotype_RIL</span><span class="p">,</span><span class="n">response</span><span class="p">,</span><span class="n">WARNING</span><span class="p">,</span><span class="n">GT_Final</span><span class="p">]</span>

                <span class="k">elif</span> <span class="p">(</span><span class="n">start</span> <span class="o">&gt;</span> <span class="n">offspring_EMS</span><span class="p">[</span><span class="n">EMS_key</span><span class="p">][</span><span class="mi">1</span><span class="p">]</span> <span class="ow">and</span> <span class="n">stop</span> <span class="o">&gt;</span> <span class="n">offspring_EMS</span><span class="p">[</span><span class="n">EMS_key</span><span class="p">][</span><span class="mi">1</span><span class="p">]):</span>
                    <span class="k">if</span> <span class="n">candidate</span> <span class="ow">in</span> <span class="n">info_RIL</span> <span class="ow">and</span> <span class="n">info_RIL</span><span class="p">[</span><span class="n">candidate</span><span class="p">][:</span><span class="mi">2</span><span class="p">]</span> <span class="o">==</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">]:</span>
                        <span class="n">genotype_RIL</span> <span class="o">=</span> <span class="n">info_RIL</span><span class="p">[</span><span class="n">candidate</span><span class="p">][</span><span class="mi">18</span><span class="p">]</span>
                    <span class="n">genotype_EMS</span> <span class="o">=</span> <span class="n">offspring_EMS</span><span class="p">[</span><span class="n">EMS_key</span><span class="p">][</span><span class="mi">4</span><span class="p">]</span>
                    <span class="n">response</span> <span class="o">=</span> <span class="s2">&quot;Yes&quot;</span> <span class="k">if</span> <span class="n">genotype_EMS</span> <span class="o">==</span> <span class="s2">&quot;A&quot;</span> <span class="k">else</span> <span class="s2">&quot;No&quot;</span>

                    <span class="k">if</span> <span class="n">genotype_RIL</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col/Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;GT_NA&quot;</span>
                            <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                    <span class="k">elif</span> <span class="n">genotype_RIL</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                        <span class="k">if</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;NEG_A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;A&quot;</span>                      <span class="c1"># modif</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Col/Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;FALSE_A&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>   <span class="c1"># modif</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;RIL&quot;</span>
                        <span class="k">elif</span> <span class="n">genotype</span> <span class="o">==</span> <span class="s2">&quot;Ct&quot;</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">response</span> <span class="o">==</span> <span class="s2">&quot;Yes&quot;</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;IMPO&quot;</span>
                                <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;GT_NA&quot;</span>
                            <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">WARNING</span> <span class="o">=</span> <span class="s2">&quot;RIL_NA&quot;</span>
                        <span class="n">GT_Final</span> <span class="o">=</span> <span class="s2">&quot;NA&quot;</span>

                    <span class="n">RULEs</span><span class="p">[</span><span class="n">candidate</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span><span class="n">stop</span><span class="p">,</span><span class="n">snp_info</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span><span class="n">snp_info</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span><span class="n">genotype</span><span class="p">,</span> <span class="n">genotype_EMS</span><span class="p">,</span> <span class="n">genotype_RIL</span><span class="p">,</span><span class="n">response</span><span class="p">,</span><span class="n">WARNING</span><span class="p">,</span><span class="n">GT_Final</span><span class="p">]</span>

                <span class="nb">chr</span> <span class="o">=</span> <span class="n">candidate</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>
                <span class="n">win_id</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">candidate</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>
                <span class="k">if</span> <span class="nb">chr</span> <span class="ow">in</span> <span class="n">nb_window_chr</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">win_id</span> <span class="o">&gt;</span> <span class="n">nb_window_chr</span><span class="p">[</span><span class="nb">chr</span><span class="p">]:</span>
                        <span class="n">nb_window_chr</span><span class="p">[</span><span class="nb">chr</span><span class="p">]</span> <span class="o">=</span> <span class="n">win_id</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">nb_window_chr</span><span class="p">[</span><span class="nb">chr</span><span class="p">]</span> <span class="o">=</span> <span class="n">win_id</span>
    <span class="k">return</span> <span class="n">RULEs</span><span class="p">,</span> <span class="n">nb_window_chr</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>Rules_RIls</cite>:</p>
<ul>
<li><p><strong>Input</strong> :
-info_RIL : A dictionary containing the genotype information for recombinant inbred lines (RILs), indexed by window identifiers. Each entry includes the start and stop positions and the RIL genotype.
-offspring_EMS : A dictionary containing the genotype data from an EMS-treated offspring dataset, with keys as window identifiers and values including start, stop positions, and genotype details.
-offspring_basic : Similar to <cite>offspring_EMS</cite> but holds genotype data for a basic, possibly untreated offspring dataset.</p></li>
<li><p><strong>Functionality</strong> :
-Cross-referencing Genotypes: The function processes three sets of genotype data by comparing genotypes from two offspring datasets (EMS and basic) with RIL genotypes to find overlaps and discrepancies within the genomic windows.
-Genotype Decision Making:</p>
<blockquote>
<div><ul class="simple">
<li><p>Evaluates overlaps between the genomic windows of the basic offspring data and both EMS offspring and RIL data.</p></li>
<li><p>For each overlapping window, it assesses the genetic data to establish a final genotype by applying predefined rules based on the consistency and discrepancy among the observed genotypes.</p></li>
<li><p>Assigns a final decision and a warning for potential issues or discrepancies such as unexpected genotype combinations or mismatches with expected RIL genotypes.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><dl class="simple">
<dt>Tracking and Logging:</dt><dd><ul>
<li><p>Collects and logs detailed information about each evaluated window, including the start and stop positions, SNP information from the basic dataset, the evaluated genotypes, and the derived conclusions.</p></li>
<li><p>Maintains a count of the number of windows per chromosome which aids in further genomic analyses and references.</p></li>
</ul>
</dd>
</dl>
</li>
</ul>
</li>
<li><p><strong>Output</strong> :
- RULEs : This dictionary contains the results for each genomic window from the basic offspring dataset. Each entry lists the genomic window’s start and stop positions, SNP info, final genotype decision, and any warnings about genotype discrepancies or unexpected findings.
- nb_window_chr : A dictionary that tracks the highest window number for each chromosome, which is useful for understanding the distribution and density of genomic data across chromosomes.</p></li>
</ul>
<p>The function also allows for the adjustment of parameters to set the threshold of support needed to validate a crossover, as well as the required size to validate a genotype.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="config_rules_RIL.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">config_rules_RIL.yaml</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="Test_yaml_detectCOs_BC_EMS.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">Test_yaml_detectCOs_BC_EMS.yaml</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Mohamad Yassine
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">detectCOs_identifyCOs_RIL.py</a><ul>
<li><a class="reference internal" href="#function-identifycos-rils">Function : IdentifyCOs_RILs</a></li>
<li><a class="reference internal" href="#function-rules-rils">Function : Rules_RIls</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>