<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="detectCOs_required_functions_EMS.py" href="detectCOs_required_functions_EMS.html" /><link rel="prev" title="VisualiseGenotypesChromosome.py" href="VisualiseGenotypesChromosome.html" />

    <!-- Generated with Sphinx 5.3.0 and Furo 2023.03.27 -->
        <title>detectCOs_identifyCOs_EMS.py - detectCOs 2.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=fad236701ea90a88636c2a8c73b44ae642ed2a53" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">detectCOs 2.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">detectCOs 2.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="config_detectCOs_CG.html">config_detectCOs_CG.yaml</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_required_functions.html">detectCOs_required_functions.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_read_files.html">detectCOs_read_files.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_sliding_window.html">detectCOs_sliding_window.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_identifyCOs.html">detectCOs_identifyCOs.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="test_detectCOs_polyrec_functions.html">test_detectCOs_polyrec_functions.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="Test_laucher_With_diff_SNPs.html">Test_laucher_With_diff_SNPs.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="Test_yaml_detectCOs_BC.html">Test_yaml_detectCOs_BC.yaml</a></li>
<li class="toctree-l1"><a class="reference internal" href="VisualiseGenotypesChromosome.html">VisualiseGenotypesChromosome.py</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">detectCOs_identifyCOs_EMS.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_required_functions_EMS.html">detectCOs_required_functions_EMS.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_sliding_window_EMS.html">detectCOs_sliding_window_EMS.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="Test_yaml_detectCOs_BC_EMS.html">Test_yaml_detectCOs_BC_EMS.yaml</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="detectcos-identifycos-ems-py">
<h1>detectCOs_identifyCOs_EMS.py<a class="headerlink" href="#detectcos-identifycos-ems-py" title="Permalink to this heading">#</a></h1>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">hues</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<span class="kn">from</span> <span class="nn">detectCOs_required_functions_EMS</span> <span class="kn">import</span> <span class="o">*</span>
</pre></div>
</div>
<section id="function-identifycos-snps">
<h2>Function : IdentifyCOs_snps<a class="headerlink" href="#function-identifycos-snps" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">IdentifyCOs_snps</span><span class="p">(</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">,</span> <span class="n">nb_windows_chr</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:    Find location of hypothetical cross-over according to the change of genotype</span>

<span class="sd">    Input:                  - offspring_genotype_window_smoothed: Ordered dictionnary</span>
<span class="sd">                    from SmoothNormalizedOsffspringSlidingWindow containing :</span>
<span class="sd">                    [chr_window] = [start, stop, ADref/DP, ADalt/DP, probHomoRef, probHetero, probHomoAlt, genotype]</span>

<span class="sd">                    - nb_windows_chr:       number of window per chromosome, determine with function</span>
<span class="sd">                    NormalizeOffspringSlidingWindow()</span>

<span class="sd">    Output:                 Cross-over candidate from smoothed windows</span>
<span class="sd">                    candidates_co[chr_mean_window] = [left_window_border + &quot;:&quot; + right_window_border,</span>
<span class="sd">                                                    co pos start, co pos stop, previous genotype, new genotype]</span>

<span class="sd">                    All double cross-over found during the analysis are saved, even if they are under the threshold</span>
<span class="sd">                    to support the genotype (thr_support) :</span>
<span class="sd">                    db_co[chr_mean_window] = [cur_chr, start_co_win:stop_co_win, previous genotype, new genotype, status]</span>

<span class="sd">    Upgrade? :      add input parameter : length required to support genotype (kb, default 300kb) and window size (kb)</span>
<span class="sd">                modify :</span>
<span class="sd">                    if stop_win &lt;= 10 or start_win &gt; (nb_win - 10):</span>
<span class="sd">                        thr_support = int((thr_support_kb/window_size) - 1) # (thr_support_kb/2)/(window_size/2) - 1</span>
<span class="sd">                    else:</span>
<span class="sd">                        thr_support = int((thr_support_kb/(window_size/2)) - 1)</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="nb">print</span><span class="p">()</span>
    <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Identify candidate crossovers&quot;</span><span class="p">)</span>
    <span class="c1">#log(&quot;Identify candidate crossovers&quot;)</span>

    <span class="n">candidates_co</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">db_co</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span> <span class="c1"># save all homo to another homo genotype even if they are not enough supported</span>
    <span class="n">thr_support</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># number of windows to support genotype = threshold support</span>

    <span class="k">for</span> <span class="n">cur_chr</span><span class="p">,</span> <span class="n">nb_win</span> <span class="ow">in</span> <span class="n">nb_windows_chr</span><span class="o">.</span><span class="n">items</span><span class="p">():</span> <span class="c1">## iteration on each chromosome</span>
        <span class="c1">#log(f&quot;\nProcessing chromosome: {cur_chr} with {nb_win} windows.&quot;)</span>
        <span class="n">nb_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">nb_win</span><span class="p">)</span>

        <span class="c1"># information about the window num_win</span>
        <span class="n">cur_support</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># number of window that support current genotype</span>
        <span class="n">pre_num</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># number of window that support current genotype after a NA transition # modif MY</span>
        <span class="n">cur_geno</span> <span class="o">=</span> <span class="nb">str</span><span class="p">()</span> <span class="c1"># genotype of current window(=geno_stop)</span>
        <span class="n">cur_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># current window = num_win + 1</span>

        <span class="c1"># information about the previous supported geonotype</span>
        <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># genotype of the window before the hypothetical COs</span>
        <span class="n">pre_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># last window of the previous genotype (before hypothetical COs)</span>

        <span class="c1"># count between previous genotype and current genotype</span>
        <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># number of consecutive NA window</span>
        <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># support of genotype before starting NA region</span>
        <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># initialize geno_preNA at the start of processing each chromosome # modify MY</span>


        <span class="k">for</span> <span class="n">num_win</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">nb_win</span><span class="p">):</span> <span class="c1"># iteration on the each window of each chromosome cur_chr</span>
            <span class="c1">#log(f&quot;\nAnalyzing window {num_win}/{nb_win} on chromosome {cur_chr}.&quot;)</span>
            <span class="c1"># Define genotype for window start and stop</span>
            <span class="n">start_win</span> <span class="o">=</span> <span class="n">num_win</span> <span class="c1"># from 1 to the penultimate num_win (due to range function)</span>
            <span class="n">key_start</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">start_win</span><span class="p">)</span>
            <span class="n">geno_start</span> <span class="o">=</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">key_start</span><span class="p">][</span><span class="mi">4</span><span class="p">]</span>

            <span class="n">stop_win</span> <span class="o">=</span> <span class="n">num_win</span> <span class="o">+</span> <span class="mi">1</span> <span class="c1"># from 2 to the last num_win</span>
            <span class="n">key_stop</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_win</span><span class="p">)</span>
            <span class="n">geno_stop</span> <span class="o">=</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">key_stop</span><span class="p">][</span><span class="mi">4</span><span class="p">]</span>
            <span class="c1">## Edit the thr_support = min number of window to validate a COs</span>
            <span class="c1"># Handling the first 6 windows</span>
            <span class="k">if</span> <span class="n">start_win</span> <span class="o">&lt;=</span> <span class="mi">6</span><span class="p">:</span>
                <span class="c1"># Specific case for the 7th window</span>
                <span class="k">if</span> <span class="n">stop_win</span> <span class="o">==</span> <span class="mi">7</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">5</span>  <span class="c1"># Case where start and stop genotypes are the same and not &quot;NA&quot;</span>
                <span class="c1"># Handling the first 3 windows</span>
                <span class="k">if</span> <span class="n">start_win</span> <span class="o">==</span> <span class="mi">4</span> <span class="ow">and</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="mi">3</span><span class="p">)][</span><span class="mi">4</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">1</span>
                <span class="k">elif</span> <span class="n">start_win</span> <span class="o">&lt;=</span> <span class="mi">3</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># cases in the first 3 windows</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1"># Windows between the 4th and the 6th</span>

            <span class="c1"># Handling the last 6 windows</span>
            <span class="k">elif</span> <span class="n">start_win</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">6</span><span class="p">):</span>
                <span class="c1"># Specific case where the current window is at least (nb_win-6) but not the penultimate</span>
                <span class="k">if</span> <span class="n">start_win</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">6</span><span class="p">)</span> <span class="ow">and</span> <span class="n">start_win</span> <span class="o">&lt;=</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">6</span><span class="p">)][</span><span class="mi">4</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">5</span>  <span class="c1"># Start and stop genotypes of the window are identical, matching a previously smoothed value, not &quot;NA&quot;</span>
                <span class="k">elif</span> <span class="n">start_win</span> <span class="o">==</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">6</span><span class="p">)</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">!=</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">5</span>  <span class="c1"># Specific to the (nb_win - 6) window with differing start and stop genotypes and not &quot;NA&quot;</span>
                <span class="k">elif</span> <span class="n">start_win</span> <span class="o">&gt;=</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">3</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">start_win</span> <span class="o">==</span> <span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">3</span><span class="p">)</span> <span class="ow">and</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)][</span><span class="mi">4</span><span class="p">]</span> <span class="o">==</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">nb_win</span> <span class="o">-</span> <span class="mi">3</span><span class="p">)][</span><span class="mi">4</span><span class="p">]</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                        <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">2</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># The last three windows with differing end genotypes from the previous smoothing and not &quot;NA&quot;</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">2</span>  <span class="c1"># Other cases in the last 6 windows</span>

            <span class="c1"># Cases for all other windows not specified above</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">5</span>  <span class="c1"># Default value for windows that are neither among the first nor the last 6</span>

            <span class="c1">#log(f&quot;Window start genotype: {geno_start}, Window stop genotype: {geno_stop}&quot;)</span>

            <span class="c1">#log(f&quot;Current threshold support: {thr_support}.&quot;)</span>
            <span class="c1">## Initialization of the start and stop posistion of the CO</span>
            <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># mean position of the start window of the crossover</span>
            <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># mean position of the stop window of the crossover</span>
            <span class="n">stop_co_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># number of window of the end of the crossover region</span>


            <span class="k">if</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="n">geno_stop</span> <span class="p">:</span>
                <span class="c1">#log(&quot;Genotype change detected between consecutive windows.&quot;)</span>

                <span class="k">if</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">or</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="c1">#log(&quot;Handling NA genotype scenario...&quot;)</span>
                    <span class="k">if</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="p">:</span>
                        <span class="c1">#log(&quot;Handling geno_stop == &#39;NA&#39; and  geno_start != &#39;NA&#39; scenario...&quot;)</span>
                        <span class="c1"># when enter in NA region:</span>
                        <span class="n">geno_preNA</span> <span class="o">=</span> <span class="n">geno_start</span>
                        <span class="n">win_preNA</span> <span class="o">=</span> <span class="n">start_win</span>
                        <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">1</span>
                        <span class="n">support_preNA</span> <span class="o">=</span> <span class="n">cur_support</span>  <span class="c1">#modify MY</span>
                        <span class="c1">#log(f&#39;geno_preNA = {geno_preNA}, win_preNA = {win_preNA}, count_NA = {count_NA}, support_preNA = {support_preNA}&#39;)</span>
                        <span class="k">continue</span>

                    <span class="k">elif</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="n">geno_preNA</span> <span class="p">:</span>
                        <span class="c1">#log(f&quot;Handling geno_start == &#39;NA&#39; and  geno_stop ({geno_stop}) == geno_preNA ({geno_preNA}) scenario...&quot;)</span>
                        <span class="c1">#log(f&#39;pre_win = {pre_win}, pre_geno = {pre_geno}&#39;)</span>
                        <span class="n">cur_support</span> <span class="o">+=</span><span class="mi">1</span>
                        <span class="c1">#log(f&#39;cur_support = {cur_support}&#39;)</span>

                        <span class="n">support_preNA</span> <span class="o">=</span> <span class="n">cur_support</span>  <span class="c1"># modify MY</span>
                        <span class="c1">#log(f&#39;support_preNA = {support_preNA}&#39;)</span>
                        <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&gt;=</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">!=</span>  <span class="n">geno_stop</span><span class="p">:</span>      <span class="c1">## voir cas 13   # modif MY</span>
                            <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>
                            <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
                            <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>

                            <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                            <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                            <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                            <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                            <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                            <span class="c1">#log(f&quot;key {key} supported by {cur_support} in candidateCO.&quot;)</span>
                            <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

                            <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                                <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>

                            <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># modify MY</span>
                            <span class="n">pre_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span>                           <span class="c1"># test MY</span>
                            <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># modify MY</span>
                            <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># modify MY</span>
                            <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>  <span class="c1"># Also reset geno_preNA  # modify MY</span>

                        <span class="k">continue</span>

                    <span class="k">elif</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">!=</span> <span class="n">geno_preNA</span><span class="p">:</span>
                        <span class="c1">#log(f&quot;Handling geno_start == &#39;NA&#39; and  geno_stop ({geno_stop}) != geno_preNA ({geno_preNA}) scenario...&quot;)</span>
                        <span class="k">if</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>
                            <span class="c1"># if geno_stop != geno_preNA, edit pre_geno and pre_win with the information before NA region</span>
                            <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">geno_preNA</span> <span class="c1"># genotype before NA region</span>
                            <span class="n">pre_win</span> <span class="o">=</span> <span class="n">win_preNA</span> <span class="c1"># last window before NA region</span>
                            <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                            <span class="n">win_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="k">else</span><span class="p">:</span> <span class="c1"># Added by MY</span>
                            <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">pre_geno</span>  <span class="c1"># test MY</span>
                            <span class="n">pre_win</span> <span class="o">=</span> <span class="n">pre_win</span> <span class="c1"># test MY</span>
                            <span class="n">cur_support</span> <span class="o">=</span> <span class="n">pre_num</span> <span class="c1"># test MY</span>
                            <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span> <span class="c1"># test MY</span>
                            <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span> <span class="c1"># test MY</span>
                            <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># test MY</span>
                            <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># test MY</span>
                            <span class="k">continue</span> <span class="c1"># test MY</span>

                <span class="k">elif</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>
                    <span class="c1">#log(f&quot;cur_support = {cur_support}, in the first side.&quot;)</span>
                    <span class="c1"># if genotype change between start and stop num_win and</span>
                    <span class="c1"># genotype supported by at least threshold support (2 or 5)</span>
                    <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">geno_start</span> <span class="c1"># genotype of window before CO</span>
                    <span class="n">pre_win</span> <span class="o">=</span> <span class="n">start_win</span> <span class="c1"># window before CO</span>
                    <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># modify MY</span>
                    <span class="n">pre_num</span> <span class="o">=</span> <span class="n">cur_support</span> <span class="c1"># useless</span>

                <span class="n">cur_support</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># reset cur_support</span>

                <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">geno_start</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">geno_stop</span><span class="p">)</span> <span class="ow">and</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="n">key_db_co</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">((</span><span class="n">start_win</span> <span class="o">+</span> <span class="n">stop_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                    <span class="n">db_co</span><span class="p">[</span><span class="n">key_db_co</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">start_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_win</span><span class="p">),</span> <span class="n">geno_start</span><span class="p">,</span> <span class="n">geno_stop</span><span class="p">,</span> <span class="s2">&quot;not_supported&quot;</span><span class="p">]</span>
                    <span class="c1">#log(f&quot;We are here {key_db_co}.&quot;)</span>

                <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>  <span class="c1"># Added by MY</span>
                    <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
                    <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>
                    <span class="c1">#log(&quot;Crossover candidate detected due to genotype change and validate thr_support in last window.&quot;)</span>
                    <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&lt;</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">support_preNA</span> <span class="o">!=</span> <span class="mi">0</span> <span class="p">:</span>
                        <span class="c1"># if there is some NA window during the count of cur_support</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="k">else</span> <span class="p">:</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">+</span> <span class="mi">1</span>

                    <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                    <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                    <span class="c1">## Edit candidate</span>
                    <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                    <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>
                    <span class="c1">#log(f&quot;key {key} supported by {cur_support} in candidateCO at last window.&quot;)</span>

                    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                        <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>

                    <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">pre_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span>                           <span class="c1"># test MY</span>

            <span class="k">else</span><span class="p">:</span> <span class="c1"># geno_start == geno_stop</span>
                <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
                <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>

                <span class="k">if</span> <span class="n">cur_geno</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="p">:</span>
                    <span class="n">count_NA</span> <span class="o">+=</span> <span class="mi">1</span>
                    <span class="k">continue</span>

                <span class="n">cur_support</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="c1">#log(f&quot;Genotype {geno_stop} supported by {cur_support} consecutive windows.&quot;)</span>
                <span class="c1">#log(f&quot;pre_geno {pre_geno}.&quot;)</span>
                <span class="c1">#log(f&quot;cur_geno {cur_geno}.&quot;)</span>
                <span class="c1">#log(f&quot;thr_support {thr_support}.&quot;)</span>

                <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">cur_geno</span><span class="p">:</span>
                    <span class="k">continue</span>

                <span class="k">elif</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span> <span class="c1"># modify MY</span>
                    <span class="c1">#log(&quot;Crossover candidate detected due to genotype change and validate thr_support.&quot;)</span>
                    <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&lt;</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">support_preNA</span> <span class="o">!=</span> <span class="mi">0</span> <span class="p">:</span>
                        <span class="c1"># if there is some NA window during the count of cur_support</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>
                    <span class="k">else</span> <span class="p">:</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">+</span> <span class="mi">1</span>

                    <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                    <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                    <span class="c1">## Edit candidate</span>
                    <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                    <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>
                    <span class="c1">#log(f&quot;key {key} supported by {cur_support} in candidateCO without NA.&quot;)</span>

                    <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                        <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>

                    <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="c1">### Raise warning for all double COs found in the data</span>
    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">db_co</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Be careful, some double COs has been found during the analysis :&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">db_co</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1"># value = [cur_chr, start_co_win:stop_co_win, previous genotype, new genotype, status]</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- &quot;</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="s2">&quot; (window &quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;): &quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> \
                <span class="s2">&quot; (&quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="s2">&quot;)&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>
            <span class="c1">#log(&quot;- &quot; + key + &quot; (window &quot; + value[1] + &quot;): &quot; + value[2] + &quot;&gt;&quot; + value[3] + &quot; (&quot; + value[4] + &quot;)&quot;)</span>

    <span class="k">return</span> <span class="n">candidates_co</span><span class="p">,</span> <span class="n">db_co</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>IdentifyCOs_snps</cite>: This function searches for the locations of hypothetical crossover events based on changes in genotype.</p>
<ul class="simple">
<li><p><strong>Input:</strong>
- <cite>offspring_genotype_window_smoothed</cite>: An ordered dictionary produced by the <cite>SmoothNormalizedOffspringSlidingWindow</cite> function containing smoothed genotype data of the offspring.
- <cite>nb_windows_chr</cite>: The number of windows per chromosome, determined using the <cite>NormalizeOffspringSlidingWindow</cite> function.</p></li>
<li><p><strong>Output:</strong>
- <cite>candidates_co</cite>: Candidates for crossovers, derived from the smoothed windows.
- <cite>db_co</cite>: All double crossovers found during analysis, even if they are not sufficiently supported.</p></li>
<li><p><strong>Functionality:</strong>
- The function iterates through each chromosome and examines each genotype data window.
- It detects changes in genotype between windows and identifies the start and end positions of the crossovers.
- Crossovers are recorded with their average positions and the genotypes before and after the crossover.
- Double crossovers are also recorded, even if they do not have sufficient support.</p></li>
</ul>
<p>The function also supports parameters to adjust the threshold of support needed to validate a crossover, as well as the required size to support a genotype.</p>
</section>
<section id="function-precisecos-snps">
<h2>Function : PreciseCOs_snps<a class="headerlink" href="#function-precisecos-snps" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">PreciseCOs_snps</span><span class="p">(</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">,</span> <span class="n">start_window</span><span class="o">=</span><span class="kc">None</span><span class="p">,</span> <span class="n">end_window</span><span class="o">=</span><span class="kc">None</span><span class="p">):</span>   <span class="c1"># By MY</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Find location of hypothetical cross-over according to the change of genotype</span>

<span class="sd">    :param offspring_genotype_window_smoothed: Ordered dictionary from SmoothNormalizedOsffspringSlidingWindow containing :</span>
<span class="sd">                        [chr_window] = [start, stop, ADref/DP, ADalt/DP, probHomoRef, probHetero, probHomoAlt, genotype]</span>
<span class="sd">    :param nb_windows_chr: Number of window per chromosome, determined with function NormalizeOffspringSlidingWindow()</span>
<span class="sd">    :param start_window: Start window to begin analysis (inclusive)</span>
<span class="sd">    :param end_window: End window to stop analysis (inclusive)</span>
<span class="sd">    :return: Cross-over candidate from smoothed windows and all double cross-over found during the analysis</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="c1"># Initial setup</span>
    <span class="n">candidates_co</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">db_co</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">thr_support</span> <span class="o">=</span> <span class="mi">3</span>
    <span class="c1"># If start_window or end_window are not specified, use defaults</span>
    <span class="k">if</span> <span class="n">start_window</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">start_win</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">start_win</span> <span class="o">=</span> <span class="n">start_window</span>
    <span class="k">if</span> <span class="n">end_window</span> <span class="ow">is</span> <span class="kc">None</span><span class="p">:</span>
        <span class="n">end_window</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="nb">list</span><span class="p">(</span><span class="n">offspring_genotype_window_smoothed</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">1</span><span class="p">])</span>

    <span class="n">cur_chr</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">offspring_genotype_window_smoothed</span><span class="o">.</span><span class="n">keys</span><span class="p">())[</span><span class="o">-</span><span class="mi">1</span><span class="p">]</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)[</span><span class="mi">0</span><span class="p">]</span>


    <span class="c1"># Initial variables for tracking changes</span>
    <span class="n">pre_num</span> <span class="o">=</span> <span class="mi">1</span>  <span class="c1"># number of window that support current genotype after a NA transition # modif MY</span>
    <span class="n">cur_support</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="n">cur_win</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span>
    <span class="n">pre_geno</span><span class="p">,</span> <span class="n">pre_win</span> <span class="o">=</span> <span class="s1">&#39;&#39;</span><span class="p">,</span> <span class="mi">0</span>
    <span class="n">count_NA</span><span class="p">,</span> <span class="n">support_preNA</span><span class="p">,</span> <span class="n">geno_preNA</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="s1">&#39;&#39;</span>

    <span class="k">for</span> <span class="n">num_win</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">start_win</span><span class="p">,</span> <span class="n">end_window</span><span class="p">):</span>  <span class="c1"># Adjusted to work within specified window range</span>

        <span class="c1"># Define genotype for window start and stop</span>
        <span class="n">start_win</span> <span class="o">=</span> <span class="n">num_win</span> <span class="c1"># from 1 to the penultimate num_win (due to range function)</span>
        <span class="n">key_start</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">start_win</span><span class="p">)</span>
        <span class="n">geno_start</span> <span class="o">=</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">key_start</span><span class="p">][</span><span class="mi">4</span><span class="p">]</span>

        <span class="n">stop_win</span> <span class="o">=</span> <span class="n">num_win</span> <span class="o">+</span> <span class="mi">1</span> <span class="c1"># from 2 to the last num_win</span>
        <span class="n">key_stop</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_win</span><span class="p">)</span>
        <span class="n">geno_stop</span> <span class="o">=</span> <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">key_stop</span><span class="p">][</span><span class="mi">4</span><span class="p">]</span>
        <span class="c1">#(f&quot;Window start genotype: {geno_start}, Window stop genotype: {geno_stop}&quot;)</span>

        <span class="n">relative_position</span> <span class="o">=</span> <span class="n">num_win</span> <span class="o">-</span> <span class="n">start_window</span> <span class="o">+</span> <span class="mi">1</span>
        <span class="n">relative_end_position</span> <span class="o">=</span> <span class="n">end_window</span> <span class="o">-</span> <span class="n">start_window</span> <span class="o">+</span> <span class="mi">1</span>

        <span class="c1">## Initialization of the start and stop posistion of the CO</span>
        <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># mean position of the start window of the crossover</span>
        <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># mean position of the stop window of the crossover</span>
        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="nb">int</span><span class="p">()</span> <span class="c1"># number of window of the end of the crossover region</span>


        <span class="k">if</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="n">geno_stop</span> <span class="p">:</span>
            <span class="c1">#print(&quot;Genotype change detected between consecutive windows.&quot;)</span>

            <span class="k">if</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">or</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                <span class="c1">#print(&quot;Handling NA genotype scenario...&quot;)</span>
                <span class="k">if</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="p">:</span>
                    <span class="c1">#print(&quot;Handling NA genotype scenario...&quot;)</span>

                    <span class="c1"># when enter in NA region:</span>
                    <span class="n">geno_preNA</span> <span class="o">=</span> <span class="n">geno_start</span>
                    <span class="n">win_preNA</span> <span class="o">=</span> <span class="n">start_win</span>
                    <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="n">support_preNA</span> <span class="o">=</span> <span class="n">cur_support</span>

                    <span class="k">continue</span>

                <span class="k">elif</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">==</span> <span class="n">geno_preNA</span> <span class="p">:</span>
                    <span class="n">cur_support</span> <span class="o">+=</span><span class="mi">1</span>
                    <span class="n">support_preNA</span> <span class="o">=</span> <span class="n">cur_support</span>  <span class="c1"># modify MY</span>

                    <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&gt;=</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">!=</span>  <span class="n">geno_stop</span><span class="p">:</span>  <span class="c1">## voir cas 13  # modif MY</span>
                        <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>
                        <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
                        <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>

                        <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                        <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                        <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                        <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                        <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

                        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                            <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>
                        <span class="c1"># Reset variables after detecting a CO</span>
                        <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># modify MY</span>
                        <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>  <span class="c1"># modify MY</span>
                        <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># modify MY</span>
                        <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>  <span class="c1"># Also reset geno_preNA  # modify MY</span>

                    <span class="k">continue</span>

                <span class="k">elif</span> <span class="n">geno_start</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">!=</span> <span class="n">geno_preNA</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>
                        <span class="c1"># Logic to record crossover event</span>
                        <span class="c1"># if geno_stop != geno_preNA, edit pre_geno and pre_win with the information before NA region</span>
                        <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">geno_preNA</span> <span class="c1"># genotype before NA region</span>
                        <span class="n">pre_win</span> <span class="o">=</span> <span class="n">win_preNA</span> <span class="c1"># last window before NA region</span>
                        <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">win_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="k">else</span><span class="p">:</span> <span class="c1"># Added by MY</span>
                        <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">pre_geno</span>  <span class="c1"># test MY</span>
                        <span class="n">pre_win</span> <span class="o">=</span> <span class="n">pre_win</span> <span class="c1"># test MY</span>
                        <span class="n">cur_support</span> <span class="o">=</span> <span class="n">pre_num</span> <span class="c1"># test MY</span>
                        <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span> <span class="c1"># test MY</span>
                        <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span> <span class="c1"># test MY</span>
                        <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># test MY</span>
                        <span class="n">geno_preNA</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span> <span class="c1"># test MY</span>
                        <span class="k">continue</span> <span class="c1"># test MY</span>

            <span class="k">elif</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>
                <span class="c1"># if genotype change between start and stop num_win and</span>
                <span class="c1"># genotype supported by at least threshold support (2 or 5)</span>
                <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">geno_start</span> <span class="c1"># genotype of window before CO</span>
                <span class="n">pre_win</span> <span class="o">=</span> <span class="n">start_win</span> <span class="c1"># window before CO</span>
                <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span> <span class="c1"># modify MY</span>
                <span class="n">pre_num</span> <span class="o">=</span> <span class="n">cur_support</span> <span class="c1"># useless</span>

            <span class="n">cur_support</span> <span class="o">=</span> <span class="mi">1</span> <span class="c1"># reset cur_support</span>

            <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">geno_start</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">geno_stop</span><span class="p">)</span> <span class="ow">and</span> <span class="n">geno_start</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno_stop</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                <span class="n">key_db_co</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">((</span><span class="n">start_win</span> <span class="o">+</span> <span class="n">stop_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>
                <span class="n">db_co</span><span class="p">[</span><span class="n">key_db_co</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">start_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_win</span><span class="p">),</span> <span class="n">geno_start</span><span class="p">,</span> <span class="n">geno_stop</span><span class="p">,</span> <span class="s2">&quot;not_supported&quot;</span><span class="p">]</span>
            <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>  <span class="c1"># Added by MY</span>
                <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
                <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>
                <span class="c1">#log(&quot;Crossover candidate detected due to genotype change and validate thr_support in last window.&quot;)</span>
                <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&lt;</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">support_preNA</span> <span class="o">!=</span> <span class="mi">0</span> <span class="p">:</span>
                    <span class="c1"># if there is some NA window during the count of cur_support</span>
                    <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="k">else</span> <span class="p">:</span>
                    <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">+</span> <span class="mi">1</span>

                <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                    <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                    <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                <span class="c1">## Edit candidate</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>
                <span class="c1">#log(f&quot;key {key} supported by {cur_support} in candidateCO at last window.&quot;)</span>

                <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                    <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>

                <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="k">else</span><span class="p">:</span> <span class="c1"># geno_start == geno_stop</span>
            <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno_stop</span>
            <span class="n">cur_win</span> <span class="o">=</span> <span class="n">stop_win</span>

            <span class="k">if</span> <span class="n">cur_geno</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="p">:</span>
                <span class="n">count_NA</span> <span class="o">+=</span> <span class="mi">1</span>
                <span class="k">continue</span>

            <span class="n">cur_support</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="c1">#print(f&quot;Genotype {geno_stop} supported by {cur_support} consecutive windows.&quot;)</span>

            <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">cur_geno</span><span class="p">:</span>
                <span class="k">continue</span>

            <span class="k">elif</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">cur_support</span> <span class="o">&gt;=</span> <span class="n">thr_support</span><span class="p">:</span>
                <span class="c1">#print(&quot;Crossover candidate detected due to genotype change.&quot;)</span>
                <span class="k">if</span> <span class="n">support_preNA</span> <span class="o">&lt;</span> <span class="n">thr_support</span> <span class="ow">and</span> <span class="n">support_preNA</span> <span class="o">!=</span> <span class="mi">0</span> <span class="p">:</span>
                    <span class="c1"># if there is some NA window during the count of cur_support</span>
                    <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">-</span> <span class="n">count_NA</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="k">else</span> <span class="p">:</span>
                    <span class="n">stop_co_win</span> <span class="o">=</span> <span class="n">cur_win</span> <span class="o">-</span> <span class="n">cur_support</span> <span class="o">+</span> <span class="mi">1</span>

                <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                    <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">((</span><span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> \
                    <span class="n">offspring_genotype_window_smoothed</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span><span class="p">)</span>

                <span class="c1">## Edit candidate</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_win</span> <span class="o">+</span> <span class="n">stop_co_win</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                <span class="n">candidates_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

                <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="p">:</span>
                    <span class="n">db_co</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_win</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;:&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop_co_win</span><span class="p">),</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">,</span> <span class="s2">&quot;supported&quot;</span><span class="p">]</span>

                <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="n">count_NA</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">support_preNA</span> <span class="o">=</span> <span class="mi">0</span>


    <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">db_co</span><span class="p">)</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Be careful, some double COs has been found during the analysis :&quot;</span><span class="p">)</span>
        <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">db_co</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1"># value = [cur_chr, start_co_win:stop_co_win, previous genotype, new genotype, status]</span>
            <span class="nb">print</span><span class="p">(</span><span class="s2">&quot;- &quot;</span><span class="p">,</span> <span class="n">key</span><span class="p">,</span> <span class="s2">&quot; (window &quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">1</span><span class="p">],</span><span class="s2">&quot;): &quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">2</span><span class="p">],</span> <span class="s2">&quot;&gt;&quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">3</span><span class="p">],</span> <span class="s2">&quot; (&quot;</span><span class="p">,</span> <span class="n">value</span><span class="p">[</span><span class="mi">4</span><span class="p">],</span> <span class="s2">&quot;)&quot;</span><span class="p">,</span> <span class="n">sep</span><span class="o">=</span><span class="s2">&quot;&quot;</span><span class="p">)</span>


    <span class="k">return</span> <span class="n">candidates_co</span><span class="p">,</span> <span class="n">db_co</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>PreciseCOs_snps</cite>: This function identifies the locations of potential genetic crossovers by analyzing changes in genotype.</p>
<ul class="simple">
<li><p><strong>Input:</strong>
- <cite>offspring_genotype_window_smoothed</cite>: An ordered dictionary from the <cite>SmoothNormalizedOffspringSlidingWindow</cite> function, containing smoothed genotype data of the offspring adjusted for the new window size.
- <cite>start_window</cite>: The starting coordinate for the analysis, based on data obtained in <cite>candidateCO</cite>.
- <cite>end_window</cite>: The stopping coordinate for the analysis, also based on data obtained in <cite>candidateCO</cite>.</p></li>
<li><p><strong>Output:</strong>
- <cite>candidates_co</cite>: Candidates for crossovers, identified from the smoothed data windows.
- <cite>db_co</cite>: All double crossovers detected during the analysis, even if they are not sufficiently supported.</p></li>
<li><p><strong>Functionality:</strong>
- The function iterates through each chromosome and examines the genotype data windows between predefined start and stop coordinates.
- It detects changes in genotype between the windows and identifies the start and end positions of the crossovers.
- Crossovers are recorded with their average positions and the genotypes before and after the crossover.
- Double crossovers are also recorded, even if they do not have sufficient support.</p></li>
</ul>
<p>The function also allows for the adjustment of parameters to set the threshold of support needed to validate a crossover, as well as the required size to validate a genotype.</p>
</section>
<section id="function-refinecoborders">
<h2>Function : RefineCOBorders<a class="headerlink" href="#function-refinecoborders" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">RefineCOBorders</span><span class="p">(</span><span class="n">CandidateCOs</span><span class="p">,</span> <span class="n">Offspring_infoSNPs</span><span class="p">,</span> <span class="n">window_size</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">()</span>
    <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Refine crossover borders&quot;</span><span class="p">)</span>

    <span class="n">RefinedCOs</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">co_key</span><span class="p">,</span> <span class="n">co_value</span> <span class="ow">in</span> <span class="n">CandidateCOs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

        <span class="n">cur_chr</span><span class="p">,</span> <span class="n">cur_pos</span> <span class="o">=</span> <span class="n">co_key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="n">win</span><span class="p">,</span> <span class="n">cur_start</span><span class="p">,</span> <span class="n">cur_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">co_value</span>

        <span class="n">cur_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
        <span class="n">cur_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">):</span>

            <span class="n">cur_geno1</span><span class="p">,</span> <span class="n">cur_geno2</span> <span class="o">=</span> <span class="n">cur_geno</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)</span>
            <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="n">pre_geno</span>
            <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

            <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">cur_geno1</span><span class="p">:</span>
                <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="n">cur_geno2</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="n">cur_geno1</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="n">pre_geno</span>

            <span class="n">refined_infosnp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_infosnp_cnt</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">refined_infosnp_pos</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cur_start</span><span class="p">,</span> <span class="n">cur_stop</span><span class="p">):</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">Offspring_infoSNPs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">snp_geno</span> <span class="o">=</span> <span class="n">Offspring_infoSNPs</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">5</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">pre_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">cur_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">cur_start</span>
            <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">cur_stop</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">refined_infosnp_cnt</span><span class="p">):</span>

                <span class="n">snp_pos</span><span class="p">,</span> <span class="n">snp_type</span> <span class="o">=</span> <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&gt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                    <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">snp_pos</span>
                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&lt;</span> <span class="n">refined_info_stop</span><span class="p">:</span>
                    <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">snp_pos</span>
                    <span class="k">if</span> <span class="n">refined_info_start</span> <span class="o">!=</span> <span class="n">cur_start</span><span class="p">:</span>
                        <span class="k">break</span>

            <span class="k">if</span> <span class="n">refined_info_stop</span> <span class="o">&lt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">cur_start</span> <span class="o">-</span> <span class="mi">2000</span>
                <span class="c1"># hues.warn(&quot;ERROR-info-2, stop &lt; start!&quot;)</span>

            <span class="n">cur_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">cur_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="c1"># hues.info(&quot;#CandidateCOs-2: &quot; + str(cur_start) + &quot;, &quot; + str(cur_stop) + &quot;, &quot; + pre_geno + &quot;, &quot; + cur_geno)</span>
            <span class="c1"># hues.info(&quot;#RefinedInfoCOs-2: &quot; + str(refined_info_start) + &quot;, &quot; + str(refined_info_stop) + &quot;, &quot; + pre_geno + &quot;, &quot; + cur_geno)</span>
            <span class="n">RefinedCOs</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">refined_info_start</span><span class="p">,</span> <span class="n">refined_info_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

        <span class="k">elif</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">):</span>

            <span class="n">pre_geno1</span><span class="p">,</span> <span class="n">pre_geno2</span> <span class="o">=</span> <span class="n">pre_geno</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)</span>
            <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="n">cur_geno</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

            <span class="k">if</span> <span class="n">pre_geno1</span> <span class="o">==</span> <span class="n">cur_geno</span><span class="p">:</span>
                <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="n">pre_geno2</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="n">pre_geno1</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="n">pre_geno1</span>

            <span class="n">refined_infosnp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_infosnp_cnt</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">refined_infosnp_pos</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cur_start</span><span class="p">,</span> <span class="n">cur_stop</span><span class="p">):</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">Offspring_infoSNPs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">snp_geno</span> <span class="o">=</span> <span class="n">Offspring_infoSNPs</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">5</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">pre_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">cur_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">cur_start</span>
            <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">cur_stop</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">refined_infosnp_cnt</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span>

                <span class="n">snp_pos</span><span class="p">,</span> <span class="n">snp_type</span> <span class="o">=</span> <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&gt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                    <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">snp_pos</span>

                    <span class="k">if</span> <span class="n">refined_info_stop</span> <span class="o">!=</span> <span class="n">cur_stop</span><span class="p">:</span>
                        <span class="k">break</span>

                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&lt;</span> <span class="n">refined_info_stop</span><span class="p">:</span>
                    <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">snp_pos</span>

            <span class="k">if</span> <span class="n">refined_info_stop</span> <span class="o">&lt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">cur_stop</span> <span class="o">+</span> <span class="mi">2000</span>
                <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;ERROR-info-3, stop &lt; start!&quot;</span><span class="p">)</span>

            <span class="n">cur_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">cur_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="c1"># hues.info(&quot;#CandidateCOs-3: &quot; + str(cur_start) + &quot;, &quot; + str(cur_stop) + &quot;, &quot; + pre_geno + &quot;, &quot; + cur_geno)</span>
            <span class="c1"># hues.info(&quot;#RefinedInfoCOs-3: &quot; + str(refined_info_start) + &quot;, &quot; + str(refined_info_stop) + &quot;, &quot; + pre_geno + &quot;, &quot; + cur_geno)</span>
            <span class="n">RefinedCOs</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">refined_info_start</span><span class="p">,</span> <span class="n">refined_info_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="n">pre_geno</span>
            <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="n">cur_geno</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

            <span class="n">refined_infosnp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_infosnp_cnt</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">refined_infosnp_pos</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_totalsnp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_totalsnp_cnt</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cur_start</span><span class="p">,</span> <span class="n">cur_stop</span><span class="p">):</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">Offspring_infoSNPs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">snp_geno</span> <span class="o">=</span> <span class="n">Offspring_infoSNPs</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">5</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">pre_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">cur_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">cur_start</span>
            <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">cur_stop</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">refined_infosnp_cnt</span><span class="p">):</span>

                <span class="n">snp_pos</span><span class="p">,</span> <span class="n">snp_type</span> <span class="o">=</span> <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&gt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                    <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">snp_pos</span>

                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&lt;</span> <span class="n">refined_info_stop</span><span class="p">:</span>
                    <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">snp_pos</span>

                    <span class="k">if</span> <span class="n">refined_info_start</span> <span class="o">!=</span> <span class="n">cur_start</span><span class="p">:</span>
                        <span class="k">break</span>

            <span class="k">if</span> <span class="n">refined_info_stop</span> <span class="o">&lt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                <span class="n">refined_info_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
                <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
                <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;ERROR-info-4, stop &lt; start!&quot;</span><span class="p">)</span>

            <span class="n">cur_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">cur_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;#CandidateCOs-4: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_start</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_stop</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="n">pre_geno</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="n">cur_geno</span><span class="p">)</span>
            <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;#RefinedInfoCOs-4: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">refined_info_start</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">refined_info_stop</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="n">pre_geno</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="n">cur_geno</span><span class="p">)</span>
            <span class="n">RefinedCOs</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">refined_info_start</span><span class="p">,</span> <span class="n">refined_info_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

    <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Re-refined crossovers&quot;</span><span class="p">)</span>
    <span class="n">Re_RefinedCOs</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_start</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">cur_item</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">co_key</span><span class="p">,</span> <span class="n">co_value</span> <span class="ow">in</span> <span class="n">RefinedCOs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

        <span class="n">co_chr</span><span class="p">,</span> <span class="n">co_pos</span> <span class="o">=</span> <span class="n">co_key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">co_geno1</span><span class="p">,</span> <span class="n">co_geno2</span> <span class="o">=</span> <span class="n">co_value</span>

        <span class="k">if</span> <span class="n">co_chr</span> <span class="o">!=</span> <span class="n">pre_co_chr</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">pre_co_chr</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>

            <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="n">co_chr</span>
            <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">co_pos</span><span class="p">)</span>
            <span class="n">pre_co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_start</span><span class="p">)</span>
            <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_stop</span><span class="p">)</span>
            <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="n">co_geno1</span>
            <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="n">co_geno2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">co_start</span> <span class="o">&lt;=</span> <span class="n">pre_co_stop</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">pre_co_geno2</span> <span class="o">==</span> <span class="n">co_geno1</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">pre_co_geno1</span> <span class="o">==</span> <span class="n">co_geno2</span><span class="p">:</span>
                        <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_start</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="n">co_chr</span>
                        <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">pre_co_pos</span> <span class="o">+</span> <span class="nb">float</span><span class="p">(</span><span class="n">co_pos</span><span class="p">))</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                        <span class="n">pre_co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pre_co_start</span><span class="p">)</span>
                        <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_stop</span><span class="p">)</span>
                        <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="n">pre_co_geno1</span>
                        <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="n">co_geno2</span>
                        <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">hues</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;#3#&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;Diff GENO!&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">co_start</span> <span class="o">%</span> <span class="mi">1000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">pre_co_geno1</span> <span class="ow">in</span> <span class="n">co_geno1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">pre_co_geno2</span> <span class="ow">in</span> <span class="n">co_geno1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">pre_co_geno2</span> <span class="o">==</span> <span class="n">co_geno2</span><span class="p">):</span>
                    <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="n">co_chr</span>
                    <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">pre_co_pos</span> <span class="o">+</span> <span class="nb">float</span><span class="p">(</span><span class="n">co_pos</span><span class="p">))</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="n">pre_co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pre_co_start</span><span class="p">)</span>
                    <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pre_co_stop</span><span class="p">)</span>
                    <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="n">pre_co_geno1</span>
                    <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="n">co_geno2</span>
                    <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>
                    <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="n">co_chr</span>
                    <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">co_pos</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="n">pre_co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_start</span><span class="p">)</span>
                    <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_stop</span><span class="p">)</span>
                    <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="n">co_geno1</span>
                    <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="n">co_geno2</span>

        <span class="n">cur_item</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">cur_item</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">RefinedCOs</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">Re_RefinedCOs</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>RefineCOBorders</cite>: This function refines the borders of genetic crossovers that were previously detected. It takes three parameters as input: <cite>CandidateCOs</cite>, <cite>Offspring_infoSNPs</cite>, and <cite>window_size</cite>.
- <cite>CandidateCOs</cite>: An ordered dictionary containing candidates for crossovers. Each entry in the dictionary is a key-value pair, where the key is a string representing the chromosome and position, and the value is a list containing the details of the crossover candidate.
- <cite>Offspring_infoSNPs</cite>: A dictionary containing information about the SNPs (Single Nucleotide Polymorphisms) of the offspring.
- <cite>window_size</cite>: The size of the window used in the analysis.</p>
<p>The function begins by initializing an empty dictionary called <cite>RefinedCOs</cite> to store the refined crossovers.</p>
<dl class="simple">
<dt>Next, it iterates through each crossover candidate in <cite>CandidateCOs</cite> and performs the following operations:</dt><dd><ul class="simple">
<li><p>Extract the details of the crossover candidate.</p></li>
<li><p>Adjust the start and end positions of the crossover based on the window size.</p></li>
<li><dl class="simple">
<dt>For each SNP between the start and end positions of the crossover:</dt><dd><ul>
<li><p>If the genotype matches the previous genotype (before the crossover), record the SNPs position.</p></li>
</ul>
</dd>
</dl>
</li>
<li><p>Determine the new start and end positions of the crossover based on the found SNPs.</p></li>
<li><p>Record the refined crossover in the <cite>RefinedCOs</cite> dictionary.</p></li>
</ul>
</dd>
</dl>
<p>The function then returns the <cite>RefinedCOs</cite> dictionary containing the refined crossovers.</p>
</section>
<section id="function-identifycosqichao">
<h2>Function : IdentifyCOsQichao<a class="headerlink" href="#function-identifycosqichao" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">IdentifyCOsQichao</span><span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">,</span> <span class="n">Offspring_smoothWinNums</span><span class="p">,</span> \
                    <span class="n">Offspring_slidingGenoNums</span><span class="p">,</span> <span class="n">Offspring_slidingGenoRatios</span><span class="p">,</span> \
                    <span class="n">Centromere</span><span class="p">,</span> <span class="n">genoRef</span><span class="p">,</span> <span class="n">genoAlt</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">()</span>
    <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Identify candidate crossovers&quot;</span><span class="p">)</span>

    <span class="n">CandidateCOs</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>

    <span class="n">window</span> <span class="o">=</span> <span class="mi">2</span>

    <span class="k">for</span> <span class="n">cur_chr</span><span class="p">,</span> <span class="n">win_num</span> <span class="ow">in</span> <span class="n">Offspring_smoothWinNums</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
        <span class="n">min_win_num</span> <span class="o">=</span> <span class="mi">5</span>
        <span class="n">max_win_num</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">win_num</span><span class="p">)</span>

        <span class="n">cur_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">cur_pos</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">cur_num</span> <span class="o">=</span> <span class="mi">1</span>
        <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">pre_pos</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">pre_num</span> <span class="o">=</span> <span class="mi">0</span>

        <span class="n">cur1_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
        <span class="n">cur1_pos1</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">cur1_pos2</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="n">cur1_num</span> <span class="o">=</span> <span class="mi">1</span>

        <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">max_win_num</span><span class="p">):</span>

            <span class="n">start</span> <span class="o">=</span> <span class="n">x</span>
            <span class="n">stop</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">window</span> <span class="o">-</span> <span class="mi">1</span> <span class="c1"># start+1!</span>

            <span class="k">if</span> <span class="n">stop</span> <span class="o">&gt;</span> <span class="n">win_num</span><span class="p">:</span>
                <span class="n">stop</span> <span class="o">=</span> <span class="n">win_num</span>

            <span class="k">if</span> <span class="n">stop</span> <span class="o">&lt;</span> <span class="mi">10</span> <span class="ow">or</span> <span class="n">start</span> <span class="o">&gt;</span> <span class="p">(</span><span class="n">win_num</span> <span class="o">-</span> <span class="mi">10</span><span class="p">):</span>
                <span class="n">min_win_num</span> <span class="o">=</span> <span class="mi">2</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">min_win_num</span> <span class="o">=</span> <span class="mi">5</span>

            <span class="n">key1</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">start</span><span class="p">)</span>
            <span class="n">geno1</span> <span class="o">=</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">key1</span><span class="p">][</span><span class="mi">5</span><span class="p">]</span>
            <span class="n">key2</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop</span><span class="p">)</span>
            <span class="n">geno2</span> <span class="o">=</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">key2</span><span class="p">][</span><span class="mi">5</span><span class="p">]</span>

            <span class="k">if</span> <span class="n">geno1</span> <span class="o">==</span> <span class="n">geno2</span><span class="p">:</span> <span class="c1">#geno</span>
                <span class="k">if</span> <span class="n">geno1</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">or</span> <span class="n">geno2</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span> <span class="c1"># should not happen ?</span>
                    <span class="k">continue</span>

                <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno2</span>
                <span class="n">cur_pos</span> <span class="o">=</span> <span class="n">stop</span>
                <span class="n">cur_num</span> <span class="o">+=</span> <span class="mi">1</span>

                <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span> <span class="c1"># pre_geno only defined when geno1!=geno2 and cur_num &gt; (min_win_num - 1)</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">or</span> <span class="n">cur_geno</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span> <span class="c1"># pre_geno only defined when geno1!=geno2 and cur_num &gt; (min_win_num - 1)</span>
                        <span class="k">continue</span>
                    <span class="k">else</span><span class="p">:</span> <span class="c1"># when a change in genotype is finally detected between geno1</span>
                        <span class="k">if</span> <span class="n">cur_num</span> <span class="o">&lt;</span> <span class="n">min_win_num</span> <span class="ow">or</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">cur_geno</span><span class="p">:</span>
                            <span class="k">if</span> <span class="n">cur_num</span> <span class="o">&gt;</span> <span class="mi">1</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">):</span>
                                <span class="n">cur1_geno</span> <span class="o">=</span> <span class="n">cur_geno</span>
                                <span class="n">cur1_pos1</span> <span class="o">=</span> <span class="n">cur_pos</span> <span class="o">-</span> <span class="n">cur_num</span> <span class="o">+</span> <span class="mi">1</span>
                                <span class="n">cur1_pos2</span> <span class="o">=</span> <span class="n">cur_pos</span>
                                <span class="n">cur1_num</span> <span class="o">=</span> <span class="n">cur_num</span>
                            <span class="k">if</span> <span class="n">cur1_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">cur_geno</span> <span class="ow">and</span> <span class="n">cur_num</span> <span class="o">&gt;=</span> <span class="n">min_win_num</span><span class="p">:</span>
                                <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;Close double COs! -- same homo&quot;</span><span class="p">)</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="k">continue</span>

                        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="n">cur1_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                            <span class="c1"># hues.warn(&quot;Close double COs!&quot;)</span>

                            <span class="k">if</span> <span class="n">cur1_pos1</span> <span class="o">&lt;</span> <span class="n">pre_pos</span> <span class="ow">or</span> <span class="n">cur1_pos2</span> <span class="o">&gt;</span> <span class="n">cur_pos</span><span class="p">:</span>
                                <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="k">if</span> <span class="n">cur1_num</span> <span class="o">&lt;</span> <span class="mi">4</span><span class="p">:</span>
                                    <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="n">cur_geno</span><span class="p">:</span>
                                        <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                        <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur1_pos1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur1_pos1</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                        <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_pos</span> <span class="o">+</span> <span class="n">cur1_pos1</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                                        <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">CandidateCOs</span><span class="p">:</span>
                                            <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur1_geno</span><span class="p">]</span>

                                        <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur1_pos2</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur1_pos2</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                        <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                        <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">cur1_pos2</span> <span class="o">+</span> <span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                                        <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">CandidateCOs</span><span class="p">:</span>
                                            <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">cur1_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>
                                    <span class="k">else</span><span class="p">:</span>
                                        <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                        <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                <span class="k">else</span><span class="p">:</span> <span class="c1"># cur1_num = 4</span>
                                    <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                    <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur1_pos1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur1_pos1</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                    <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_pos</span> <span class="o">+</span> <span class="n">cur1_pos1</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                                    <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">CandidateCOs</span><span class="p">:</span>
                                        <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur1_geno</span><span class="p">]</span>

                                    <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur1_pos2</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur1_pos2</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                    <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                                    <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">cur1_pos2</span> <span class="o">+</span> <span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                                    <span class="k">if</span> <span class="n">key</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">CandidateCOs</span><span class="p">:</span>
                                        <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">cur1_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

                            <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                            <span class="n">pre_pos</span> <span class="o">=</span> <span class="mi">0</span>
                            <span class="n">pre_num</span> <span class="o">=</span> <span class="mi">0</span>

                            <span class="k">continue</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                            <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_smoothProbs</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="n">min_win_num</span> <span class="o">+</span> <span class="mi">1</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>

                        <span class="n">cur1_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>


                        <span class="k">if</span> <span class="n">co_stop</span> <span class="o">-</span> <span class="n">co_start</span> <span class="o">&gt;</span> <span class="mi">300000</span><span class="p">:</span>
                            <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;large breakpoint interval, correct it!&quot;</span><span class="p">)</span>

                            <span class="n">cor_pre_pos</span> <span class="o">=</span> <span class="n">pre_pos</span> <span class="o">-</span> <span class="mi">10</span>
                            <span class="k">if</span> <span class="n">cor_pre_pos</span> <span class="o">&lt;</span> <span class="mi">1</span><span class="p">:</span>
                                <span class="n">cor_pre_pos</span> <span class="o">=</span> <span class="mi">1</span>

                            <span class="n">cor_cur_pos</span> <span class="o">=</span> <span class="n">cur_pos</span> <span class="o">+</span> <span class="mi">8</span>
                            <span class="k">if</span> <span class="n">cor_cur_pos</span> <span class="o">&gt;</span> <span class="n">win_num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">:</span>
                                <span class="n">cor_cur_pos</span> <span class="o">=</span> <span class="n">win_num</span> <span class="o">-</span> <span class="mi">1</span>

                            <span class="n">rawCO</span> <span class="o">=</span> <span class="p">[</span><span class="n">cur_chr</span><span class="p">,</span> <span class="n">cor_pre_pos</span><span class="p">,</span> <span class="n">cor_cur_pos</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>
                            <span class="n">correctCO</span> <span class="o">=</span> <span class="n">CorrectLargeCOsQichao</span><span class="p">(</span><span class="n">Offspring_slidingGenoNums</span><span class="p">,</span>\
                                        <span class="n">Offspring_slidingGenoRatios</span><span class="p">,</span> <span class="n">rawCO</span><span class="p">,</span> <span class="n">Centromere</span><span class="p">,</span> <span class="n">genoRef</span><span class="p">,</span> <span class="n">genoAlt</span><span class="p">)</span>

                            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">correctCO</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
                                <span class="k">for</span> <span class="n">key</span><span class="p">,</span> <span class="n">value</span> <span class="ow">in</span> <span class="n">correctCO</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
                                    <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="n">value</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_pos</span> <span class="o">+</span> <span class="n">cur_pos</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                                <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_pos</span> <span class="o">+</span> <span class="n">cur_pos</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                            <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

                        <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_pos</span> <span class="o">=</span> <span class="mi">0</span>
                        <span class="n">pre_num</span> <span class="o">=</span> <span class="mi">0</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">cur_num</span> <span class="o">&gt;</span> <span class="p">(</span><span class="n">min_win_num</span> <span class="o">-</span> <span class="mi">1</span><span class="p">):</span>
                    <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">geno1</span>
                    <span class="n">pre_pos</span> <span class="o">=</span> <span class="n">start</span>
                    <span class="n">pre_num</span> <span class="o">=</span> <span class="n">cur_num</span>
                <span class="n">cur_num</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="k">return</span> <span class="n">CandidateCOs</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>IdentifyCOsQichao</cite>: This function identifies candidate crossovers from the smoothed probability data of offspring genotypes.</p>
<ul class="simple">
<li><p><strong>Input:</strong>
- <cite>Offspring_smoothProbs</cite>: Dictionary containing the smoothed probabilities of the offsprings genotypes.
- <cite>Offspring_smoothWinNums</cite>: Dictionary containing the number of windows for each chromosome of the offspring.
- <cite>Offspring_slidingGenoNums</cite>: Dictionary containing the sliding genotype numbers of the offspring.
- <cite>Offspring_slidingGenoRatios</cite>: Dictionary containing the ratios of sliding genotypes of the offspring.
- <cite>Centromere</cite>: Dictionary containing the locations of centromeres.
- <cite>genoRef</cite>: Reference genotype.
- <cite>genoAlt</cite>: Alternative genotype.</p></li>
<li><p><strong>Functionality:</strong>
- The function begins by initializing an empty dictionary called <cite>CandidateCOs</cite> to store candidate crossovers.
- It then iterates through each chromosome in <cite>Offspring_smoothWinNums</cite>.
- For each window in the chromosome, it examines the smoothed genotypes at the start and end points of the window.
- If the genotypes are the same, it indicates that there is no genotype change within that window. The function then updates the current genotype and position variables and continues counting.
- If a genotype change is detected, it checks several conditions and calculates the start and end positions of the candidate crossover.
- It updates the previous genotype and position variables, then records the candidate crossover in the <cite>CandidateCOs</cite> dictionary.
- If a pause interval is too large, it corrects the start and end positions of the candidate crossover using a correction function <cite>CorrectLargeCOsQichao</cite>.
- Finally, the function returns the <cite>CandidateCOs</cite> dictionary containing the identified candidate crossovers.</p></li>
</ul>
<p>This function appears to identify points where there is a change in the genotypes of the offspring, which may indicate recombination points.</p>
</section>
<section id="function-correctlargecosqichao">
<h2>Function : CorrectLargeCOsQichao<a class="headerlink" href="#function-correctlargecosqichao" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">CorrectLargeCOsQichao</span><span class="p">(</span><span class="n">Offspring_slidingGenoNums</span><span class="p">,</span> <span class="n">Offspring_slidingGenoRatios</span><span class="p">,</span> <span class="n">rawCO</span><span class="p">,</span> <span class="n">Centromere</span><span class="p">,</span> <span class="n">genoRef</span><span class="p">,</span><span class="n">genoAlt</span><span class="p">,</span> <span class="n">min_homo_freq</span><span class="o">=</span><span class="mf">0.9</span><span class="p">):</span>
    <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Correct large crossovers (centromere regions)&quot;</span><span class="p">)</span>

    <span class="n">CandidateCOs</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">CorrectedCO</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">cur_chr</span><span class="p">,</span> <span class="n">rawCO_start</span><span class="p">,</span> <span class="n">rawCO_stop</span><span class="p">,</span> <span class="n">rawCO_geno1</span><span class="p">,</span> <span class="n">rawCO_geno2</span> <span class="o">=</span> <span class="n">rawCO</span>

    <span class="n">cur_chr_centro_left</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">Centromere</span><span class="p">[</span><span class="n">cur_chr</span><span class="p">][</span><span class="mi">0</span><span class="p">])</span> <span class="o">-</span> <span class="mi">150</span> <span class="o">*</span> <span class="mi">1000</span>
    <span class="n">cur_chr_centro_right</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">Centromere</span><span class="p">[</span><span class="n">cur_chr</span><span class="p">][</span><span class="mi">1</span><span class="p">])</span> <span class="o">+</span> <span class="mi">150</span> <span class="o">*</span> <span class="mi">1000</span>

    <span class="n">window</span> <span class="o">=</span> <span class="mi">2</span>
    <span class="n">sliding</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="n">cur_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">cur_pos</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">cur_num</span> <span class="o">=</span> <span class="mi">1</span>
    <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_pos</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">pre_num</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="n">pre_na_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_na_pos</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="n">pre_na_num</span> <span class="o">=</span> <span class="mi">0</span>

    <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">rawCO_start</span><span class="p">,</span> <span class="n">rawCO_stop</span> <span class="o">+</span> <span class="mi">1</span><span class="p">):</span>

        <span class="n">start</span> <span class="o">=</span> <span class="n">x</span>
        <span class="n">stop</span> <span class="o">=</span> <span class="n">start</span> <span class="o">+</span> <span class="n">window</span> <span class="c1"># x+2</span>

        <span class="n">key1</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">start</span><span class="p">)</span> <span class="c1"># window = x</span>
        <span class="n">geno1</span> <span class="o">=</span>     <span class="n">GetGenoWindow</span><span class="p">(</span><span class="n">Offspring_slidingGenoNums</span><span class="p">[</span><span class="n">key1</span><span class="p">],</span> <span class="n">genoRef</span><span class="o">=</span><span class="n">genoRef</span><span class="p">,</span> <span class="n">genoAlt</span><span class="o">=</span><span class="n">genoAlt</span><span class="p">)[</span><span class="mi">7</span><span class="p">]</span>

        <span class="n">key2</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">stop</span> <span class="o">-</span> <span class="mi">1</span><span class="p">)</span> <span class="c1"># window = x +1</span>
        <span class="n">geno2</span> <span class="o">=</span> <span class="n">GetGenoWindow</span><span class="p">(</span><span class="n">Offspring_slidingGenoNums</span><span class="p">[</span><span class="n">key2</span><span class="p">],</span> <span class="n">genoRef</span><span class="o">=</span><span class="n">genoRef</span><span class="p">,</span> <span class="n">genoAlt</span><span class="o">=</span><span class="n">genoAlt</span><span class="p">)[</span><span class="mi">7</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">geno1</span> <span class="o">==</span> <span class="n">geno2</span><span class="p">:</span>

            <span class="k">if</span> <span class="n">geno1</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span> <span class="c1"># or geno2 == &quot;NA&quot;:</span>
                <span class="k">continue</span>

            <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">geno2</span>
            <span class="n">cur_pos</span> <span class="o">=</span> <span class="n">stop</span> <span class="o">-</span> <span class="mi">1</span>
            <span class="n">cur_num</span> <span class="o">+=</span> <span class="mi">1</span>

            <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">or</span> <span class="n">cur_geno</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                    <span class="k">continue</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">if</span> <span class="n">cur_num</span> <span class="o">&lt;</span> <span class="mi">3</span> <span class="ow">or</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">cur_geno</span><span class="p">:</span>
                        <span class="k">continue</span>

                    <span class="k">if</span> <span class="n">pre_na_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span> <span class="c1"># elif</span>
                        <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_slidingGenoRatios</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_slidingGenoRatios</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                        <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_slidingGenoRatios</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_na_pos</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_slidingGenoRatios</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_na_pos</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                        <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_pos</span> <span class="o">+</span> <span class="n">pre_na_pos</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                        <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">pre_na_geno</span><span class="p">]</span>

                        <span class="n">pre_na_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_na_pos</span> <span class="o">=</span> <span class="mi">0</span>
                        <span class="n">pre_na_num</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_slidingGenoRatios</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_slidingGenoRatios</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_pos</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                        <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span> <span class="p">(</span><span class="n">Offspring_slidingGenoRatios</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)][</span><span class="mi">0</span><span class="p">]</span> <span class="o">+</span> <span class="n">Offspring_slidingGenoRatios</span><span class="p">[</span><span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_pos</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)][</span><span class="mi">1</span><span class="p">])</span> <span class="o">/</span> <span class="mi">2</span> <span class="p">)</span>
                        <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="nb">round</span><span class="p">((</span><span class="n">pre_pos</span> <span class="o">+</span> <span class="n">cur_pos</span> <span class="o">-</span> <span class="mi">2</span><span class="p">)</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">))</span>
                        <span class="n">CandidateCOs</span><span class="p">[</span><span class="n">key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

                    <span class="n">pre_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="n">pre_pos</span> <span class="o">=</span> <span class="mi">0</span>
                    <span class="n">pre_num</span> <span class="o">=</span> <span class="mi">0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">cur_num</span> <span class="o">&gt;</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">geno1</span> <span class="o">!=</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">geno1</span>
                <span class="n">pre_pos</span> <span class="o">=</span> <span class="n">start</span>
                <span class="n">pre_num</span> <span class="o">=</span> <span class="n">cur_num</span>

            <span class="k">if</span> <span class="n">geno2</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno1</span> <span class="o">!=</span> <span class="n">pre_geno</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="n">pre_na_geno</span> <span class="o">=</span> <span class="n">geno1</span>
                <span class="n">pre_na_pos</span> <span class="o">=</span> <span class="n">start</span> <span class="o">-</span> <span class="n">cur_num</span> <span class="o">+</span> <span class="mi">1</span>
                <span class="n">pre_na_num</span> <span class="o">=</span> <span class="n">cur_num</span>

            <span class="k">if</span> <span class="n">pre_pos</span> <span class="o">&gt;</span> <span class="n">pre_na_pos</span><span class="p">:</span>
                <span class="n">pre_na_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                <span class="n">pre_na_pos</span> <span class="o">=</span> <span class="mi">0</span>
                <span class="n">pre_na_num</span> <span class="o">=</span> <span class="mi">0</span>

            <span class="k">if</span> <span class="n">geno2</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span><span class="p">:</span>
                <span class="k">continue</span>
            <span class="k">elif</span> <span class="n">geno1</span> <span class="o">==</span> <span class="s2">&quot;NA&quot;</span> <span class="ow">and</span> <span class="n">geno2</span> <span class="o">==</span> <span class="n">pre_geno</span><span class="p">:</span>
                <span class="n">cur_num</span> <span class="o">+=</span> <span class="mi">1</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cur_num</span> <span class="o">=</span> <span class="mi">1</span>

    <span class="n">keepRight</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">if</span> <span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">rawCO_geno1</span><span class="p">:</span>
        <span class="n">keepRight</span> <span class="o">=</span> <span class="kc">True</span>
    <span class="k">elif</span> <span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">rawCO_geno2</span><span class="p">:</span>
        <span class="n">keepRight</span> <span class="o">=</span> <span class="kc">False</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="n">needOneMore</span> <span class="o">=</span> <span class="kc">False</span>

    <span class="k">for</span> <span class="n">co_key</span><span class="p">,</span> <span class="n">co_value</span> <span class="ow">in</span> <span class="n">CandidateCOs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span> <span class="c1"># not created yet !!!</span>
        <span class="n">co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_value</span><span class="p">[</span><span class="mi">0</span><span class="p">])</span>
        <span class="n">co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_value</span><span class="p">[</span><span class="mi">1</span><span class="p">])</span>
        <span class="n">pre_geno</span> <span class="o">=</span> <span class="n">co_value</span><span class="p">[</span><span class="mi">2</span><span class="p">]</span>
        <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">co_value</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span>

        <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">rawCO_geno1</span> <span class="ow">and</span> <span class="n">cur_geno</span> <span class="o">==</span> <span class="n">rawCO_geno2</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">co_stop</span> <span class="o">&lt;=</span> <span class="n">cur_chr_centro_left</span> <span class="ow">or</span> <span class="n">co_start</span> <span class="o">&gt;=</span> <span class="n">cur_chr_centro_right</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">CorrectedCO</span> <span class="ow">and</span> <span class="n">needOneMore</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
                    <span class="k">break</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">CorrectedCO</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">co_value</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">needOneMore</span><span class="p">:</span>
                    <span class="n">CorrectedCO</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">co_value</span>
                    <span class="n">needOneMore</span> <span class="o">=</span> <span class="kc">False</span>
                    <span class="k">break</span>
                <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">CandidateCOs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">keepRight</span> <span class="o">==</span> <span class="kc">True</span> <span class="ow">and</span> <span class="n">needOneMore</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
                    <span class="n">CorrectedCO</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
                    <span class="n">CorrectedCO</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">co_value</span>
                <span class="k">elif</span> <span class="nb">len</span><span class="p">(</span><span class="n">CandidateCOs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">keepRight</span> <span class="o">==</span> <span class="kc">False</span> <span class="ow">and</span> <span class="n">needOneMore</span> <span class="o">==</span> <span class="kc">False</span><span class="p">:</span>
                    <span class="n">CorrectedCO</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">co_value</span>
                    <span class="k">break</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">CorrectedCO</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">co_value</span>
        <span class="k">elif</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">rawCO_geno1</span> <span class="ow">and</span> <span class="n">cur_geno</span> <span class="o">!=</span> <span class="n">rawCO_geno2</span><span class="p">:</span>
            <span class="n">CorrectedCO</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">co_value</span>
        <span class="k">elif</span> <span class="n">pre_geno</span> <span class="o">!=</span> <span class="n">rawCO_geno1</span> <span class="ow">and</span> <span class="n">cur_geno</span> <span class="o">==</span> <span class="n">rawCO_geno2</span><span class="p">:</span>
            <span class="n">CorrectedCO</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">co_value</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="nb">len</span><span class="p">(</span><span class="n">CandidateCOs</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span> <span class="o">&gt;</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">rawCO_geno2</span> <span class="ow">and</span> <span class="n">cur_geno</span> <span class="o">==</span> <span class="n">rawCO_geno1</span><span class="p">:</span>
                <span class="k">if</span> <span class="n">co_stop</span> <span class="o">&lt;=</span> <span class="n">cur_chr_centro_left</span> <span class="ow">or</span> <span class="n">co_start</span> <span class="o">&gt;=</span> <span class="n">cur_chr_centro_right</span><span class="p">:</span>
                    <span class="n">CorrectedCO</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="n">co_value</span>
                    <span class="n">needOneMore</span> <span class="o">=</span> <span class="kc">True</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="k">continue</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">continue</span>

    <span class="k">return</span> <span class="n">CorrectedCO</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>CorrectLargeCOsQichao</cite>: This function corrects candidate crossovers identified in centromeric regions. Here is a summary of its operation:</p>
<ul class="simple">
<li><p><strong>Input:</strong>
- <cite>Offspring_slidingGenoNums</cite>: Dictionary containing the sliding genotype numbers of the offspring.
- <cite>Offspring_slidingGenoRatios</cite>: Dictionary containing the ratios of sliding genotypes of the offspring.
- <cite>rawCO</cite>: List containing raw information about the candidate crossover to be corrected.
- <cite>Centromere</cite>: Dictionary containing the locations of centromeres.
- <cite>genoRef</cite>: Reference genotype.
- <cite>genoAlt</cite>: Alternative genotype.
- <cite>min_homo_freq</cite>: Minimum frequency of homozygotes.</p></li>
<li><p><strong>Functionality:</strong>
- The function begins by initializing an empty dictionary called <cite>CorrectedCO</cite> to store the corrected candidate crossovers.
- It then retrieves the raw information about the candidate crossover to be corrected.
- It determines the centromeric center positions for the current chromosome.
- The function iterates through each position within the range specified by the start and end positions of the candidate crossover.
- For each position, it retrieves genotypes from adjacent positions.
- If adjacent genotypes are identical, it updates the current genotype, position, and number variables.
- If a genotype change is detected, it handles specific cases where there are NA genotypes or a genotype change in a centromeric region.
- It then stores the corrected candidate crossovers in the <cite>CorrectedCO</cite> dictionary.</p></li>
</ul>
<p>Finally, the function returns the <cite>CorrectedCO</cite> dictionary containing the corrected candidate crossovers.</p>
</section>
<section id="function-refinecobordersqichao">
<h2>Function : RefineCOBordersQichao<a class="headerlink" href="#function-refinecobordersqichao" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">RefineCOBordersQichao</span><span class="p">(</span><span class="n">CandidateCOs</span><span class="p">,</span> <span class="n">Offspring_infoSNPs</span><span class="p">,</span> <span class="n">window_size</span><span class="p">):</span>
    <span class="nb">print</span><span class="p">()</span>
    <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Refine crossover borders&quot;</span><span class="p">)</span>

    <span class="n">RefinedCOs</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>

    <span class="k">for</span> <span class="n">co_key</span><span class="p">,</span> <span class="n">co_value</span> <span class="ow">in</span> <span class="n">CandidateCOs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

        <span class="n">cur_chr</span><span class="p">,</span> <span class="n">cur_pos</span> <span class="o">=</span> <span class="n">co_key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="n">cur_start</span><span class="p">,</span> <span class="n">cur_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span> <span class="o">=</span> <span class="n">co_value</span>

        <span class="n">cur_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
        <span class="n">cur_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>

        <span class="k">if</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">):</span>

            <span class="n">cur_geno1</span><span class="p">,</span> <span class="n">cur_geno2</span> <span class="o">=</span> <span class="n">cur_geno</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)</span>
            <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="n">pre_geno</span>
            <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

            <span class="k">if</span> <span class="n">pre_geno</span> <span class="o">==</span> <span class="n">cur_geno1</span><span class="p">:</span>
                <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="n">cur_geno2</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="n">cur_geno1</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="n">pre_geno</span>

            <span class="n">refined_infosnp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_infosnp_cnt</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">refined_infosnp_pos</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cur_start</span><span class="p">,</span> <span class="n">cur_stop</span><span class="p">):</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">Offspring_infoSNPs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">snp_geno</span> <span class="o">=</span> <span class="n">Offspring_infoSNPs</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">5</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">pre_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">cur_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">cur_start</span>
            <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">cur_stop</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">refined_infosnp_cnt</span><span class="p">):</span>

                <span class="n">snp_pos</span><span class="p">,</span> <span class="n">snp_type</span> <span class="o">=</span> <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&gt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                    <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">snp_pos</span>
                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&lt;</span> <span class="n">refined_info_stop</span><span class="p">:</span>
                    <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">snp_pos</span>
                    <span class="k">if</span> <span class="n">refined_info_start</span> <span class="o">!=</span> <span class="n">cur_start</span><span class="p">:</span>
                        <span class="k">break</span>

            <span class="k">if</span> <span class="n">refined_info_stop</span> <span class="o">&lt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">cur_start</span> <span class="o">-</span> <span class="mi">2000</span>
                <span class="c1"># hues.warn(&quot;ERROR-info-2, stop &lt; start!&quot;)</span>

            <span class="n">cur_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">cur_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="c1"># hues.info(&quot;#CandidateCOs-2: &quot; + str(cur_start) + &quot;, &quot; + str(cur_stop) + &quot;, &quot; + pre_geno + &quot;, &quot; + cur_geno)</span>
            <span class="c1"># hues.info(&quot;#RefinedInfoCOs-2: &quot; + str(refined_info_start) + &quot;, &quot; + str(refined_info_stop) + &quot;, &quot; + pre_geno + &quot;, &quot; + cur_geno)</span>
            <span class="n">RefinedCOs</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">refined_info_start</span><span class="p">,</span> <span class="n">refined_info_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

        <span class="k">elif</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">in</span> <span class="n">pre_geno</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="s2">&quot;/&quot;</span> <span class="ow">not</span> <span class="ow">in</span> <span class="n">cur_geno</span><span class="p">):</span>

            <span class="n">pre_geno1</span><span class="p">,</span> <span class="n">pre_geno2</span> <span class="o">=</span> <span class="n">pre_geno</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;/&quot;</span><span class="p">)</span>
            <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
            <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="n">cur_geno</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

            <span class="k">if</span> <span class="n">pre_geno1</span> <span class="o">==</span> <span class="n">cur_geno</span><span class="p">:</span>
                <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="n">pre_geno2</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="n">pre_geno1</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="n">pre_geno1</span>

            <span class="n">refined_infosnp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_infosnp_cnt</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">refined_infosnp_pos</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cur_start</span><span class="p">,</span> <span class="n">cur_stop</span><span class="p">):</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">Offspring_infoSNPs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">snp_geno</span> <span class="o">=</span> <span class="n">Offspring_infoSNPs</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">5</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">pre_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">cur_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">cur_start</span>
            <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">cur_stop</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">refined_infosnp_cnt</span> <span class="o">-</span> <span class="mi">1</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="o">-</span><span class="mi">1</span><span class="p">):</span>

                <span class="n">snp_pos</span><span class="p">,</span> <span class="n">snp_type</span> <span class="o">=</span> <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&gt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                    <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">snp_pos</span>

                    <span class="k">if</span> <span class="n">refined_info_stop</span> <span class="o">!=</span> <span class="n">cur_stop</span><span class="p">:</span>
                        <span class="k">break</span>

                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&lt;</span> <span class="n">refined_info_stop</span><span class="p">:</span>
                    <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">snp_pos</span>

            <span class="k">if</span> <span class="n">refined_info_stop</span> <span class="o">&lt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">cur_stop</span> <span class="o">+</span> <span class="mi">2000</span>
                <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;ERROR-info-3, stop &lt; start!&quot;</span><span class="p">)</span>

            <span class="n">cur_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">cur_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="c1"># hues.info(&quot;#CandidateCOs-3: &quot; + str(cur_start) + &quot;, &quot; + str(cur_stop) + &quot;, &quot; + pre_geno + &quot;, &quot; + cur_geno)</span>
            <span class="c1"># hues.info(&quot;#RefinedInfoCOs-3: &quot; + str(refined_info_start) + &quot;, &quot; + str(refined_info_stop) + &quot;, &quot; + pre_geno + &quot;, &quot; + cur_geno)</span>
            <span class="n">RefinedCOs</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">refined_info_start</span><span class="p">,</span> <span class="n">refined_info_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

        <span class="k">else</span><span class="p">:</span>

            <span class="n">pre_geno_3</span> <span class="o">=</span> <span class="n">pre_geno</span>
            <span class="n">cur_geno_3</span> <span class="o">=</span> <span class="n">cur_geno</span>
            <span class="n">comm_geno</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>

            <span class="n">refined_infosnp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_infosnp_cnt</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="n">refined_infosnp_pos</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_totalsnp</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
            <span class="n">refined_totalsnp_cnt</span> <span class="o">=</span> <span class="mi">1</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">cur_start</span><span class="p">,</span> <span class="n">cur_stop</span><span class="p">):</span>
                <span class="n">key</span> <span class="o">=</span> <span class="n">cur_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">i</span><span class="p">)</span>

                <span class="k">if</span> <span class="n">key</span> <span class="ow">in</span> <span class="n">Offspring_infoSNPs</span><span class="o">.</span><span class="n">keys</span><span class="p">():</span>
                    <span class="n">snp_geno</span> <span class="o">=</span> <span class="n">Offspring_infoSNPs</span><span class="p">[</span><span class="n">key</span><span class="p">][</span><span class="mi">5</span><span class="p">]</span>

                    <span class="k">if</span> <span class="n">pre_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">1</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">1</span>
                    <span class="k">if</span> <span class="n">cur_geno_3</span> <span class="o">==</span> <span class="n">snp_geno</span><span class="p">:</span>
                        <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">refined_infosnp_cnt</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">i</span><span class="p">,</span> <span class="mi">2</span><span class="p">]</span>
                        <span class="n">refined_infosnp_cnt</span> <span class="o">+=</span> <span class="mi">1</span>
                        <span class="n">refined_infosnp_pos</span><span class="p">[</span><span class="n">i</span><span class="p">]</span> <span class="o">=</span> <span class="mi">2</span>

            <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">cur_start</span>
            <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">cur_stop</span>
            <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="n">refined_infosnp_cnt</span><span class="p">):</span>

                <span class="n">snp_pos</span><span class="p">,</span> <span class="n">snp_type</span> <span class="o">=</span> <span class="n">refined_infosnp</span><span class="p">[</span><span class="n">i</span><span class="p">]</span>
                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">1</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&gt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                    <span class="n">refined_info_start</span> <span class="o">=</span> <span class="n">snp_pos</span>

                <span class="k">if</span> <span class="n">snp_type</span> <span class="o">==</span> <span class="mi">2</span> <span class="ow">and</span> <span class="n">snp_pos</span> <span class="o">&lt;</span> <span class="n">refined_info_stop</span><span class="p">:</span>
                    <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="n">snp_pos</span>

                    <span class="k">if</span> <span class="n">refined_info_start</span> <span class="o">!=</span> <span class="n">cur_start</span><span class="p">:</span>
                        <span class="k">break</span>

            <span class="k">if</span> <span class="n">refined_info_stop</span> <span class="o">&lt;</span> <span class="n">refined_info_start</span><span class="p">:</span>
                <span class="n">refined_info_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
                <span class="n">refined_info_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
                <span class="n">hues</span><span class="o">.</span><span class="n">warn</span><span class="p">(</span><span class="s2">&quot;ERROR-info-4, stop &lt; start!&quot;</span><span class="p">)</span>

            <span class="n">cur_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_start</span> <span class="o">+</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">cur_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">cur_stop</span> <span class="o">-</span> <span class="n">window_size</span> <span class="o">/</span> <span class="mi">2</span> <span class="o">*</span> <span class="mi">1000</span><span class="p">)</span>
            <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;#CandidateCOs-4: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_start</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">cur_stop</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="n">pre_geno</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="n">cur_geno</span><span class="p">)</span>
            <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;#RefinedInfoCOs-4: &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">refined_info_start</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">refined_info_stop</span><span class="p">)</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="n">pre_geno</span> <span class="o">+</span> <span class="s2">&quot;, &quot;</span> <span class="o">+</span> <span class="n">cur_geno</span><span class="p">)</span>
            <span class="n">RefinedCOs</span><span class="p">[</span><span class="n">co_key</span><span class="p">]</span> <span class="o">=</span> <span class="p">[</span><span class="n">refined_info_start</span><span class="p">,</span> <span class="n">refined_info_stop</span><span class="p">,</span> <span class="n">pre_geno</span><span class="p">,</span> <span class="n">cur_geno</span><span class="p">]</span>

    <span class="n">hues</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Re-refined crossovers&quot;</span><span class="p">)</span>
    <span class="n">Re_RefinedCOs</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">()</span>
    <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_start</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
    <span class="n">cur_item</span> <span class="o">=</span> <span class="mi">0</span>
    <span class="k">for</span> <span class="n">co_key</span><span class="p">,</span> <span class="n">co_value</span> <span class="ow">in</span> <span class="n">RefinedCOs</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>

        <span class="n">co_chr</span><span class="p">,</span> <span class="n">co_pos</span> <span class="o">=</span> <span class="n">co_key</span><span class="o">.</span><span class="n">split</span><span class="p">(</span><span class="s2">&quot;_&quot;</span><span class="p">)</span>
        <span class="n">co_start</span><span class="p">,</span> <span class="n">co_stop</span><span class="p">,</span> <span class="n">co_geno1</span><span class="p">,</span> <span class="n">co_geno2</span> <span class="o">=</span> <span class="n">co_value</span>

        <span class="k">if</span> <span class="n">co_chr</span> <span class="o">!=</span> <span class="n">pre_co_chr</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">pre_co_chr</span> <span class="o">==</span> <span class="s2">&quot;&quot;</span><span class="p">:</span>
                <span class="k">pass</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>

            <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="n">co_chr</span>
            <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="n">co_pos</span><span class="p">)</span>
            <span class="n">pre_co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_start</span><span class="p">)</span>
            <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_stop</span><span class="p">)</span>
            <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="n">co_geno1</span>
            <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="n">co_geno2</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="k">if</span> <span class="n">co_start</span> <span class="o">&lt;=</span> <span class="n">pre_co_stop</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">pre_co_geno2</span> <span class="o">==</span> <span class="n">co_geno1</span><span class="p">):</span>
                    <span class="k">if</span> <span class="n">pre_co_geno1</span> <span class="o">==</span> <span class="n">co_geno2</span><span class="p">:</span>
                        <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_start</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                        <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="s2">&quot;&quot;</span>
                    <span class="k">else</span><span class="p">:</span>
                        <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="n">co_chr</span>
                        <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">pre_co_pos</span> <span class="o">+</span> <span class="nb">float</span><span class="p">(</span><span class="n">co_pos</span><span class="p">))</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                        <span class="n">pre_co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pre_co_start</span><span class="p">)</span>
                        <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_stop</span><span class="p">)</span>
                        <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="n">pre_co_geno1</span>
                        <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="n">co_geno2</span>
                        <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">hues</span><span class="o">.</span><span class="n">error</span><span class="p">(</span><span class="s2">&quot;#3#&quot;</span> <span class="o">+</span> <span class="s2">&quot;</span><span class="se">\t</span><span class="s2">&quot;</span> <span class="o">+</span> <span class="s2">&quot;Diff GENO!&quot;</span><span class="p">)</span>
            <span class="k">else</span><span class="p">:</span>
                <span class="k">if</span> <span class="p">(</span><span class="n">co_start</span> <span class="o">%</span> <span class="mi">1000</span> <span class="o">==</span> <span class="mi">0</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">pre_co_geno1</span> <span class="ow">in</span> <span class="n">co_geno1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">pre_co_geno2</span> <span class="ow">in</span> <span class="n">co_geno1</span><span class="p">)</span> <span class="ow">and</span> <span class="p">(</span><span class="n">pre_co_geno2</span> <span class="o">==</span> <span class="n">co_geno2</span><span class="p">):</span>
                    <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="n">co_chr</span>
                    <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="nb">round</span><span class="p">((</span><span class="n">pre_co_pos</span> <span class="o">+</span> <span class="nb">float</span><span class="p">(</span><span class="n">co_pos</span><span class="p">))</span> <span class="o">/</span> <span class="mi">2</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="n">pre_co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pre_co_start</span><span class="p">)</span>
                    <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">pre_co_stop</span><span class="p">)</span>
                    <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="n">pre_co_geno1</span>
                    <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="n">co_geno2</span>
                    <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>
                    <span class="n">pre_co_chr</span> <span class="o">=</span> <span class="n">co_chr</span>
                    <span class="n">pre_co_pos</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="nb">float</span><span class="p">(</span><span class="n">co_pos</span><span class="p">),</span> <span class="mi">1</span><span class="p">)</span>
                    <span class="n">pre_co_start</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_start</span><span class="p">)</span>
                    <span class="n">pre_co_stop</span> <span class="o">=</span> <span class="nb">int</span><span class="p">(</span><span class="n">co_stop</span><span class="p">)</span>
                    <span class="n">pre_co_geno1</span> <span class="o">=</span> <span class="n">co_geno1</span>
                    <span class="n">pre_co_geno2</span> <span class="o">=</span> <span class="n">co_geno2</span>

        <span class="n">cur_item</span> <span class="o">+=</span> <span class="mi">1</span>
        <span class="k">if</span> <span class="n">cur_item</span> <span class="o">==</span> <span class="nb">len</span><span class="p">(</span><span class="n">RefinedCOs</span><span class="o">.</span><span class="n">keys</span><span class="p">()):</span>
            <span class="n">Re_RefinedCOs</span><span class="p">[</span><span class="n">pre_co_chr</span> <span class="o">+</span> <span class="s2">&quot;_&quot;</span> <span class="o">+</span> <span class="nb">str</span><span class="p">(</span><span class="n">pre_co_pos</span><span class="p">)]</span> <span class="o">=</span> <span class="p">[</span><span class="n">pre_co_start</span><span class="p">,</span> <span class="n">pre_co_stop</span><span class="p">,</span> <span class="n">pre_co_geno1</span><span class="p">,</span> <span class="n">pre_co_geno2</span><span class="p">]</span>

    <span class="k">return</span> <span class="n">Re_RefinedCOs</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>RefineCOBordersQichao</cite>: This function refines the borders of candidate crossovers using additional information from SNPs.
- <strong>Input:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><cite>CandidateCOs</cite>: Dictionary containing the candidate crossovers to be refined.</p></li>
<li><p><cite>Offspring_infoSNPs</cite>: Dictionary containing information about the SNPs of the offspring.</p></li>
<li><p><cite>window_size</cite>: The size of the window used to refine the crossovers.</p></li>
</ul>
</div></blockquote>
<ul>
<li><p><strong>Functionality:</strong>
- The function starts by initializing an empty dictionary called <cite>RefinedCOs</cite> to store the refined candidate crossovers.
- It then iterates through each candidate crossover in the <cite>CandidateCOs</cite> dictionary.
- For each crossover, it extracts necessary information such as the chromosome, start and end positions, and the genotypes before and after the crossover.
- Based on the genotypes before and after the crossover, it handles three different cases:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>If the genotype before is homozygous and the genotype after is heterozygous, it searches for SNPs that match the genotype before and uses them to refine the start and end positions of the crossover.</p></li>
<li><p>If the genotype before is heterozygous and the genotype after is homozygous, it searches for SNPs that match the genotype after and uses them to refine the start and end positions of the crossover.</p></li>
<li><p>If both the genotypes before and after are heterozygous, it searches for SNPs corresponding to these two genotypes and uses them to refine the start and end positions of the crossover.</p></li>
</ol>
</div></blockquote>
<ul class="simple">
<li><p>Finally, it stores the refined candidate crossovers in the <cite>RefinedCOs</cite> dictionary.</p></li>
</ul>
</li>
</ul>
<p>After refining the candidate crossovers, the function moves to the re-refining step. It iterates through each refined crossover in <cite>RefinedCOs</cite> and merges adjacent crossovers on the same chromosome if they overlap or are very close. It also considers the genotypes to merge the crossovers correctly.</p>
<p>Finally, it returns the <cite>Re_RefinedCOs</cite> dictionary containing the refined and merged candidate crossovers.</p>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="detectCOs_required_functions_EMS.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">detectCOs_required_functions_EMS.py</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="VisualiseGenotypesChromosome.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">VisualiseGenotypesChromosome.py</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Mohamad Yassine
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">detectCOs_identifyCOs_EMS.py</a><ul>
<li><a class="reference internal" href="#function-identifycos-snps">Function : IdentifyCOs_snps</a></li>
<li><a class="reference internal" href="#function-precisecos-snps">Function : PreciseCOs_snps</a></li>
<li><a class="reference internal" href="#function-refinecoborders">Function : RefineCOBorders</a></li>
<li><a class="reference internal" href="#function-identifycosqichao">Function : IdentifyCOsQichao</a></li>
<li><a class="reference internal" href="#function-correctlargecosqichao">Function : CorrectLargeCOsQichao</a></li>
<li><a class="reference internal" href="#function-refinecobordersqichao">Function : RefineCOBordersQichao</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>