<!doctype html>
<html class="no-js" lang="en">
  <head><meta charset="utf-8"/>
    <meta name="viewport" content="width=device-width,initial-scale=1"/>
    <meta name="color-scheme" content="light dark"><meta name="generator" content="Docutils 0.19: https://docutils.sourceforge.io/" />
<link rel="index" title="Index" href="genindex.html" /><link rel="search" title="Search" href="search.html" /><link rel="next" title="detectCOs_read_files.py" href="detectCOs_read_files.html" /><link rel="prev" title="config_detectCOs_CG.yaml" href="config_detectCOs_CG.html" />

    <!-- Generated with Sphinx 5.3.0 and Furo 2023.03.27 -->
        <title>detectCOs_required_functions.py - detectCOs 2.0 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo.css?digest=fad236701ea90a88636c2a8c73b44ae642ed2a53" />
    <link rel="stylesheet" type="text/css" href="_static/styles/furo-extensions.css?digest=30d1aed668e5c3a91c3e3bf6a60b675221979f0e" />
    
    


<style>
  body {
    --color-code-background: #f8f8f8;
  --color-code-foreground: black;
  
  }
  @media not print {
    body[data-theme="dark"] {
      --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
    }
    @media (prefers-color-scheme: dark) {
      body:not([data-theme="light"]) {
        --color-code-background: #202020;
  --color-code-foreground: #d0d0d0;
  
      }
    }
  }
</style></head>
  <body>
    
    <script>
      document.body.dataset.theme = localStorage.getItem("theme") || "auto";
    </script>
    

<svg xmlns="http://www.w3.org/2000/svg" style="display: none;">
  <symbol id="svg-toc" viewBox="0 0 24 24">
    <title>Contents</title>
    <svg stroke="currentColor" fill="currentColor" stroke-width="0" viewBox="0 0 1024 1024">
      <path d="M408 442h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8zm-8 204c0 4.4 3.6 8 8 8h480c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8H408c-4.4 0-8 3.6-8 8v56zm504-486H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zm0 632H120c-4.4 0-8 3.6-8 8v56c0 4.4 3.6 8 8 8h784c4.4 0 8-3.6 8-8v-56c0-4.4-3.6-8-8-8zM115.4 518.9L271.7 642c5.8 4.6 14.4.5 14.4-6.9V388.9c0-7.4-8.5-11.5-14.4-6.9L115.4 505.1a8.74 8.74 0 0 0 0 13.8z"/>
    </svg>
  </symbol>
  <symbol id="svg-menu" viewBox="0 0 24 24">
    <title>Menu</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-menu">
      <line x1="3" y1="12" x2="21" y2="12"></line>
      <line x1="3" y1="6" x2="21" y2="6"></line>
      <line x1="3" y1="18" x2="21" y2="18"></line>
    </svg>
  </symbol>
  <symbol id="svg-arrow-right" viewBox="0 0 24 24">
    <title>Expand</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="2" stroke-linecap="round" stroke-linejoin="round" class="feather-chevron-right">
      <polyline points="9 18 15 12 9 6"></polyline>
    </svg>
  </symbol>
  <symbol id="svg-sun" viewBox="0 0 24 24">
    <title>Light mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="feather-sun">
      <circle cx="12" cy="12" r="5"></circle>
      <line x1="12" y1="1" x2="12" y2="3"></line>
      <line x1="12" y1="21" x2="12" y2="23"></line>
      <line x1="4.22" y1="4.22" x2="5.64" y2="5.64"></line>
      <line x1="18.36" y1="18.36" x2="19.78" y2="19.78"></line>
      <line x1="1" y1="12" x2="3" y2="12"></line>
      <line x1="21" y1="12" x2="23" y2="12"></line>
      <line x1="4.22" y1="19.78" x2="5.64" y2="18.36"></line>
      <line x1="18.36" y1="5.64" x2="19.78" y2="4.22"></line>
    </svg>
  </symbol>
  <symbol id="svg-moon" viewBox="0 0 24 24">
    <title>Dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-moon">
      <path stroke="none" d="M0 0h24v24H0z" fill="none" />
      <path d="M12 3c.132 0 .263 0 .393 0a7.5 7.5 0 0 0 7.92 12.446a9 9 0 1 1 -8.313 -12.454z" />
    </svg>
  </symbol>
  <symbol id="svg-sun-half" viewBox="0 0 24 24">
    <title>Auto light/dark mode</title>
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="none" stroke="currentColor"
      stroke-width="1.5" stroke-linecap="round" stroke-linejoin="round" class="icon-tabler-shadow">
      <path stroke="none" d="M0 0h24v24H0z" fill="none"/>
      <circle cx="12" cy="12" r="9" />
      <path d="M13 12h5" />
      <path d="M13 15h4" />
      <path d="M13 18h1" />
      <path d="M13 9h4" />
      <path d="M13 6h1" />
    </svg>
  </symbol>
</svg>

<input type="checkbox" class="sidebar-toggle" name="__navigation" id="__navigation">
<input type="checkbox" class="sidebar-toggle" name="__toc" id="__toc">
<label class="overlay sidebar-overlay" for="__navigation">
  <div class="visually-hidden">Hide navigation sidebar</div>
</label>
<label class="overlay toc-overlay" for="__toc">
  <div class="visually-hidden">Hide table of contents sidebar</div>
</label>



<div class="page">
  <header class="mobile-header">
    <div class="header-left">
      <label class="nav-overlay-icon" for="__navigation">
        <div class="visually-hidden">Toggle site navigation sidebar</div>
        <i class="icon"><svg><use href="#svg-menu"></use></svg></i>
      </label>
    </div>
    <div class="header-center">
      <a href="index.html"><div class="brand">detectCOs 2.0 documentation</div></a>
    </div>
    <div class="header-right">
      <div class="theme-toggle-container theme-toggle-header">
        <button class="theme-toggle">
          <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
          <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
          <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
          <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
        </button>
      </div>
      <label class="toc-overlay-icon toc-header-icon" for="__toc">
        <div class="visually-hidden">Toggle table of contents sidebar</div>
        <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
      </label>
    </div>
  </header>
  <aside class="sidebar-drawer">
    <div class="sidebar-container">
      
      <div class="sidebar-sticky"><a class="sidebar-brand" href="index.html">
  
  
  <span class="sidebar-brand-text">detectCOs 2.0 documentation</span>
  
</a><form class="sidebar-search-container" method="get" action="search.html" role="search">
  <input class="sidebar-search" placeholder="Search" name="q" aria-label="Search">
  <input type="hidden" name="check_keywords" value="yes">
  <input type="hidden" name="area" value="default">
</form>
<div id="searchbox"></div><div class="sidebar-scroll"><div class="sidebar-tree">
  <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1"><a class="reference internal" href="config_detectCOs_CG.html">config_detectCOs_CG.yaml</a></li>
<li class="toctree-l1 current current-page"><a class="current reference internal" href="#">detectCOs_required_functions.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_read_files.html">detectCOs_read_files.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_sliding_window.html">detectCOs_sliding_window.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="detectCOs_identifyCOs.html">detectCOs_identifyCOs.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="test_detectCOs_polyrec_functions.html">test_detectCOs_polyrec_functions.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="Test_laucher_With_diff_SNPs.html">Test_laucher_With_diff_SNPs.py</a></li>
<li class="toctree-l1"><a class="reference internal" href="Test_yaml_detectCOs_BC.html">Test_yaml_detectCOs_BC.yaml</a></li>
<li class="toctree-l1"><a class="reference internal" href="VisualiseGenotypesChromosome.html">VisualiseGenotypesChromosome.py</a></li>
</ul>

</div>
</div>

      </div>
      
    </div>
  </aside>
  <div class="main">
    <div class="content">
      <div class="article-container">
        <a href="#" class="back-to-top muted-link">
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
            <path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8v12z"></path>
          </svg>
          <span>Back to top</span>
        </a>
        <div class="content-icon-container">
          
<div class="theme-toggle-container theme-toggle-content">
            <button class="theme-toggle">
              <div class="visually-hidden">Toggle Light / Dark / Auto color theme</div>
              <svg class="theme-icon-when-auto"><use href="#svg-sun-half"></use></svg>
              <svg class="theme-icon-when-dark"><use href="#svg-moon"></use></svg>
              <svg class="theme-icon-when-light"><use href="#svg-sun"></use></svg>
            </button>
          </div>
          <label class="toc-overlay-icon toc-content-icon" for="__toc">
            <div class="visually-hidden">Toggle table of contents sidebar</div>
            <i class="icon"><svg><use href="#svg-toc"></use></svg></i>
          </label>
        </div>
        <article role="main">
          <section id="detectcos-required-functions-py">
<h1>detectCOs_required_functions.py<a class="headerlink" href="#detectcos-required-functions-py" title="Permalink to this heading">#</a></h1>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="kn">import</span> <span class="nn">math</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">from</span> <span class="nn">decimal</span> <span class="kn">import</span> <span class="o">*</span>
<span class="kn">import</span> <span class="nn">scipy</span>
</pre></div>
</div>
<section id="function-calculatehomoprobability">
<h2>Function : CalculateHomoProbability<a class="headerlink" href="#function-calculatehomoprobability" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">CalculateHomoProbability</span><span class="p">(</span><span class="n">a</span><span class="p">:</span><span class="nb">float</span><span class="p">,</span> <span class="n">b</span><span class="p">:</span><span class="nb">float</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:    calculate the probability to be homozygous a</span>
<span class="sd">    Input:                  a = ratio ADa/DP ; b = ratio ADb/DP</span>
<span class="sd">    Output:                 probability to be homozygous a</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prob</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">special</span><span class="o">.</span><span class="n">comb</span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span> <span class="o">*</span> <span class="nb">pow</span><span class="p">(</span><span class="mf">0.99</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span> <span class="o">*</span> <span class="nb">pow</span><span class="p">(</span><span class="mf">0.01</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">math</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">prob</span><span class="p">)</span> <span class="ow">or</span> <span class="n">math</span><span class="o">.</span><span class="n">isinf</span><span class="p">(</span><span class="n">prob</span><span class="p">):</span>
        <span class="n">prob</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">prob</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">prob</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>CalculateHomoProbability</cite>: This function calculates the probability of being homozygous for a specific allele using the binomial law formula, where <cite>a</cite> is the ADa/DP ratio (number of reads supporting allele a divided by sequencing depth) and <cite>b</cite> is the ADb/DP ratio (number of reads supporting allele b divided by sequencing depth).
- <strong>Input:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><cite>a</cite>: ADa/DP ratio.</p></li>
<li><p><cite>b</cite>: ADb/DP ratio.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p><strong>Functionality:</strong>
- The function uses <cite>scipy.special.comb()</cite> to calculate the binomial coefficient ((a + b) choose (a)).
- It multiplies the binomial coefficient by the probability of <cite>a</cite> successes (0.99) and <cite>b</cite> failures (0.01).
- If the result is a non-finite floating number (<cite>nan</cite> or <cite>inf</cite>), it converts it to a finite number using <cite>np.nan_to_num()</cite>.</p></li>
<li><p><strong>Output:</strong>
- <cite>prob</cite>: Probability of being homozygous for allele <cite>a</cite>.</p></li>
</ul>
<p>This function is useful for estimating the probability of homozygosity of an allele based on the AD/DP ratios of both alleles in a given SNP.</p>
</section>
<section id="function-calculateheteroprobability">
<h2>Function : CalculateHeteroProbability<a class="headerlink" href="#function-calculateheteroprobability" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">CalculateHeteroProbability</span><span class="p">(</span><span class="n">a</span><span class="p">:</span><span class="nb">float</span><span class="p">,</span> <span class="n">b</span><span class="p">:</span><span class="nb">float</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:    calculate the probability to be heterozygous</span>
<span class="sd">    Input:                  a = ratio ADa/DP ; b = ratio ADb/DP</span>
<span class="sd">    Output:                 probability to be heterozygous</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">prob</span> <span class="o">=</span> <span class="n">scipy</span><span class="o">.</span><span class="n">special</span><span class="o">.</span><span class="n">comb</span><span class="p">(</span><span class="n">a</span><span class="o">+</span><span class="n">b</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span> <span class="o">*</span> <span class="nb">pow</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">a</span><span class="p">)</span> <span class="o">*</span> <span class="nb">pow</span><span class="p">(</span><span class="mf">0.5</span><span class="p">,</span> <span class="n">b</span><span class="p">)</span>
    <span class="k">if</span> <span class="n">math</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">prob</span><span class="p">)</span> <span class="ow">or</span> <span class="n">math</span><span class="o">.</span><span class="n">isinf</span><span class="p">(</span><span class="n">prob</span><span class="p">):</span>
        <span class="n">prob</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan_to_num</span><span class="p">(</span><span class="n">prob</span><span class="p">)</span>
    <span class="k">return</span> <span class="n">prob</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>CalculateHeteroProbability</cite>: This function calculates the probability of being heterozygous for two different alleles. The probability is calculated using the binomial law formula, where <cite>a</cite> is the ADa/DP ratio (number of reads supporting allele a divided by sequencing depth) and <cite>b</cite> is the ADb/DP ratio (number of reads supporting allele b divided by sequencing depth).
- <strong>Input:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><cite>a</cite>: ADa/DP ratio.</p></li>
<li><p><cite>b</cite>: ADb/DP ratio.</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p><strong>Functionality:</strong>
- The function uses <cite>scipy.special.comb()</cite> to calculate the binomial coefficient ((a + b) choose (a)).
- It multiplies the binomial coefficient by the probability of <cite>a</cite> successes (0.5) and <cite>b</cite> failures (0.5).
- If the result is a non-finite floating number (<cite>nan</cite> or <cite>inf</cite>), it converts it to a finite number using <cite>np.nan_to_num()</cite>.</p></li>
<li><p><strong>Output:</strong>
- <cite>prob</cite>: Probability of being heterozygous.</p></li>
</ul>
<p>This function is useful for estimating the probability of heterozygosity of a SNP based on the AD/DP ratios of both alleles.</p>
</section>
<section id="function-getgenowindow">
<h2>Function : GetGenoWindow<a class="headerlink" href="#function-getgenowindow" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">GetGenoWindow</span><span class="p">(</span><span class="n">cur_window</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span> <span class="n">genoRef</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">genoAlt</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">min_homo_freq</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.9</span><span class="p">):</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:    determine the genotype according to the ratio of ADref/DP</span>
<span class="sd">                    and ADalt/DP</span>
<span class="sd">    Input:  cur_window = [start, stop, ADref, ADalt, DP]</span>
<span class="sd">    Output: cur_geno = [start, stop, ADref/DP, ADalt/DP, probHomoRef,</span>
<span class="sd">                        probHetero, probHomoAlt, genotype]</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="n">min_hetero_freq</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">min_homo_freq</span>

    <span class="c1"># Save each element of the list cur_window into specific variable</span>
    <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">ADref</span><span class="p">,</span> <span class="n">ADalt</span><span class="p">,</span> <span class="n">depth</span> <span class="o">=</span> <span class="n">cur_window</span>

    <span class="c1"># Calculate ratio ADref/DP and ADalt/DP</span>
    <span class="k">if</span> <span class="n">depth</span> <span class="o">==</span> <span class="mi">0</span> <span class="p">:</span>
        <span class="n">ratio_ADref_DP</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>
        <span class="n">ratio_ADalt_DP</span> <span class="o">=</span> <span class="nb">float</span><span class="p">(</span><span class="mf">0.0</span><span class="p">)</span>
    <span class="k">else</span> <span class="p">:</span>
        <span class="n">ratio_ADref_DP</span> <span class="o">=</span> <span class="n">ADref</span><span class="o">/</span><span class="n">depth</span>
        <span class="n">ratio_ADalt_DP</span> <span class="o">=</span> <span class="n">ADalt</span><span class="o">/</span><span class="n">depth</span>

    <span class="c1"># Define the probability to be homozygous Ref/Alt and heterozygous</span>
    <span class="n">prob_homoRef</span> <span class="o">=</span> <span class="n">prob_homoAlt</span> <span class="o">=</span> <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">0.0</span>

    <span class="k">if</span> <span class="n">ratio_ADref_DP</span> <span class="o">&gt;</span> <span class="n">ratio_ADalt_DP</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ratio_ADref_DP</span> <span class="o">&gt;=</span> <span class="n">min_homo_freq</span> <span class="ow">or</span> <span class="n">ratio_ADalt_DP</span> <span class="o">&lt;=</span> <span class="n">min_hetero_freq</span><span class="p">:</span>
            <span class="n">prob_homoRef</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">elif</span> <span class="n">ratio_ADref_DP</span> <span class="o">&gt;=</span> <span class="mf">0.3</span> <span class="ow">and</span> <span class="n">ratio_ADref_DP</span> <span class="o">&lt;=</span> <span class="mf">0.7</span><span class="p">:</span>
            <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">prob_homoRef</span> <span class="o">=</span> <span class="n">CalculateHomoProbability</span><span class="p">(</span><span class="n">ADref</span><span class="p">,</span> <span class="n">ADalt</span><span class="p">)</span>
            <span class="n">prob_hetero</span> <span class="o">=</span> <span class="n">CalculateHeteroProbability</span><span class="p">(</span><span class="n">ADref</span><span class="p">,</span> <span class="n">ADalt</span><span class="p">)</span>
            <span class="n">prob_homoAlt</span> <span class="o">=</span> <span class="n">CalculateHomoProbability</span><span class="p">(</span><span class="n">ADalt</span><span class="p">,</span> <span class="n">ADref</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">prob_homoRef</span> <span class="o">==</span> <span class="n">prob_homoAlt</span> <span class="o">==</span> <span class="n">prob_hetero</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="k">elif</span> <span class="n">ratio_ADref_DP</span> <span class="o">==</span> <span class="n">ratio_ADalt_DP</span> <span class="ow">and</span> <span class="n">ratio_ADref_DP</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="k">elif</span> <span class="n">ratio_ADref_DP</span> <span class="o">&lt;</span> <span class="n">ratio_ADalt_DP</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ratio_ADalt_DP</span> <span class="o">&gt;=</span> <span class="n">min_homo_freq</span> <span class="ow">or</span> <span class="n">ratio_ADref_DP</span> <span class="o">&lt;=</span> <span class="n">min_hetero_freq</span><span class="p">:</span>
            <span class="n">prob_homoAlt</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">elif</span> <span class="n">ratio_ADalt_DP</span> <span class="o">&gt;=</span> <span class="mf">0.3</span> <span class="ow">and</span> <span class="n">ratio_ADalt_DP</span> <span class="o">&lt;=</span> <span class="mf">0.7</span><span class="p">:</span>
            <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">prob_homoRef</span> <span class="o">=</span> <span class="n">CalculateHomoProbability</span><span class="p">(</span><span class="n">ADref</span><span class="p">,</span> <span class="n">ADalt</span><span class="p">)</span>
            <span class="n">prob_hetero</span> <span class="o">=</span> <span class="n">CalculateHeteroProbability</span><span class="p">(</span><span class="n">ADref</span><span class="p">,</span> <span class="n">ADalt</span><span class="p">)</span>
            <span class="n">prob_homoAlt</span> <span class="o">=</span> <span class="n">CalculateHomoProbability</span><span class="p">(</span><span class="n">ADalt</span><span class="p">,</span> <span class="n">ADref</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">prob_homoRef</span> <span class="o">==</span> <span class="n">prob_homoAlt</span> <span class="o">==</span> <span class="n">prob_hetero</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="k">else</span><span class="p">:</span> <span class="c1"># ratio ADref/DP == ADalt/DP == 0</span>
        <span class="c1"># do nothing: prob_homoRef = prob_homoAlt = prob_hetero = 0.0</span>
        <span class="k">pass</span>

    <span class="c1"># Edit prob_geno</span>
    <span class="n">prob_geno</span> <span class="o">=</span> <span class="p">[</span><span class="n">prob_homoRef</span><span class="p">,</span> <span class="n">prob_hetero</span><span class="p">,</span> <span class="n">prob_homoAlt</span><span class="p">,</span> <span class="s1">&#39;NA&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">prob_homoRef</span> <span class="o">&gt;</span> <span class="n">prob_hetero</span> <span class="ow">and</span> <span class="n">prob_homoRef</span> <span class="o">&gt;</span> <span class="n">prob_homoAlt</span><span class="p">:</span>
        <span class="n">prob_geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">genoRef</span>
    <span class="k">elif</span> <span class="n">prob_homoAlt</span> <span class="o">&gt;</span> <span class="n">prob_hetero</span> <span class="ow">and</span> <span class="n">prob_homoAlt</span> <span class="o">&gt;</span> <span class="n">prob_homoRef</span><span class="p">:</span>
        <span class="n">prob_geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">genoAlt</span>
    <span class="k">elif</span> <span class="n">prob_hetero</span> <span class="o">&gt;</span> <span class="n">prob_homoRef</span> <span class="ow">and</span> <span class="n">prob_hetero</span> <span class="o">&gt;</span> <span class="n">prob_homoAlt</span><span class="p">:</span>
        <span class="n">prob_geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">genoRef</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">genoAlt</span>
    <span class="k">else</span><span class="p">:</span> <span class="c1"># prob_homoRef == prob_homoAlt == heteroAB == 0.0</span>
        <span class="k">pass</span> <span class="c1"># do nothing: prob_geno = [0.0, 0.0, 0.0, &quot;NA&quot;]</span>

    <span class="n">cur_geno</span> <span class="o">=</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">ratio_ADref_DP</span><span class="p">,</span> <span class="n">ratio_ADalt_DP</span><span class="p">]</span> <span class="o">+</span> <span class="n">prob_geno</span>

    <span class="k">return</span> <span class="n">cur_geno</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>GetGenoWindow</cite>: This function determines the genotype of a genetic window based on the ratios of reads aligned to the reference and alternative alleles, as well as sequencing depth. It returns a list containing information about the genetic window, including probabilities of being homozygous for the reference and alternative alleles, the probability of being heterozygous, the predicted genotype, etc.
- <strong>Input:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><cite>cur_window</cite>: List containing elements of the window [start, stop, ADref, ADalt, DP].</p></li>
<li><p><cite>genoRef</cite>: Reference genotype.</p></li>
<li><p><cite>genoAlt</cite>: Alternative genotype.</p></li>
<li><p><cite>min_homo_freq</cite>: Minimum homozygosity frequency (defaulted to 0.9).</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p><strong>Functionality:</strong>
- The function first calculates the ADref/DP and ADalt/DP ratios.
- Then, it determines the probabilities of being homozygous for the reference genotype (<cite>prob_homoRef</cite>), homozygous for the alternative allele (<cite>prob_homoAlt</cite>), and heterozygous (<cite>prob_hetero</cite>) based on these ratios.
- If the probabilities are not clear from the ratios, they are calculated using the <cite>CalculateHomoProbability</cite> and <cite>CalculateHeteroProbability</cite> functions.
- Finally, it determines the genotype of the window based on the probabilities and returns it in <cite>cur_geno</cite>.</p></li>
<li><p><strong>Output:</strong>
- <cite>cur_geno</cite>: List containing the elements of the window along with the determined probabilities and genotype.</p></li>
</ul>
</section>
<section id="function-updated-getgenowindow">
<h2>Function : Updated GetGenoWindow<a class="headerlink" href="#function-updated-getgenowindow" title="Permalink to this heading">#</a></h2>
<p><strong>Code :</strong></p>
<div class="highlight-python notranslate"><div class="highlight"><pre><span></span><span class="k">def</span> <span class="nf">GetGenoWindow</span><span class="p">(</span><span class="n">cur_window</span><span class="p">:</span><span class="nb">list</span><span class="p">,</span> <span class="n">genoRef</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">genoAlt</span><span class="p">:</span><span class="nb">str</span><span class="p">,</span> <span class="n">min_homo_freq</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">0.75</span><span class="p">,</span> <span class="n">depth_division_th</span><span class="p">:</span><span class="nb">float</span><span class="o">=</span><span class="mf">1.0</span><span class="p">):</span>  <span class="c1"># modif MY</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">    Description:    determine the genotype according to the ratio of ADref/DP and ADalt/DP.</span>
<span class="sd">    Input:  cur_window = [start, stop, ADref, ADalt, DP, depth_division]</span>
<span class="sd">    Output: cur_geno = [start, stop, ADref/DP, ADalt/DP, genotype]</span>
<span class="sd">    &quot;&quot;&quot;</span>
    <span class="n">min_hetero_freq</span> <span class="o">=</span> <span class="mi">1</span> <span class="o">-</span> <span class="n">min_homo_freq</span>
    <span class="c1"># Save each element of the list cur_window into specific variable</span>
    <span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">ADref</span><span class="p">,</span> <span class="n">ADalt</span><span class="p">,</span> <span class="n">depth</span> <span class="o">=</span> <span class="n">cur_window</span>

    <span class="c1"># Calculate ratio ADref/DP and ADalt/DP</span>
    <span class="k">if</span> <span class="n">depth</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">ratio_ADref_DP</span> <span class="o">=</span> <span class="n">ratio_ADalt_DP</span> <span class="o">=</span> <span class="mf">0.0</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="n">ratio_ADref_DP</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">ADref</span><span class="o">/</span><span class="n">depth</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span>  <span class="c1"># modif MY</span>
        <span class="n">ratio_ADalt_DP</span> <span class="o">=</span> <span class="nb">round</span><span class="p">(</span><span class="n">ADalt</span><span class="o">/</span><span class="n">depth</span><span class="p">,</span> <span class="mi">3</span><span class="p">)</span> <span class="c1"># modif MY</span>


    <span class="c1"># Define the probability to be homozygous Ref/Alt and heterozygous</span>
    <span class="n">prob_homoRef</span> <span class="o">=</span> <span class="n">prob_homoAlt</span> <span class="o">=</span> <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">0.0</span>

    <span class="k">if</span> <span class="n">ratio_ADref_DP</span> <span class="o">&gt;</span> <span class="n">ratio_ADalt_DP</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ratio_ADref_DP</span> <span class="o">&gt;=</span> <span class="n">min_homo_freq</span> <span class="ow">or</span> <span class="n">ratio_ADalt_DP</span> <span class="o">&lt;=</span> <span class="n">min_hetero_freq</span><span class="p">:</span>
            <span class="n">prob_homoRef</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="k">elif</span> <span class="n">ratio_ADref_DP</span> <span class="o">==</span> <span class="n">ratio_ADalt_DP</span> <span class="ow">and</span> <span class="n">ratio_ADref_DP</span> <span class="o">!=</span> <span class="mi">0</span><span class="p">:</span>
        <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="k">elif</span> <span class="n">ratio_ADref_DP</span> <span class="o">&lt;</span> <span class="n">ratio_ADalt_DP</span><span class="p">:</span>
        <span class="k">if</span> <span class="n">ratio_ADalt_DP</span> <span class="o">&gt;=</span> <span class="n">min_homo_freq</span> <span class="ow">or</span> <span class="n">ratio_ADref_DP</span> <span class="o">&lt;=</span> <span class="n">min_hetero_freq</span><span class="p">:</span>
            <span class="n">prob_homoAlt</span> <span class="o">=</span> <span class="mf">1.0</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">prob_hetero</span> <span class="o">=</span> <span class="mf">1.0</span>

    <span class="k">else</span><span class="p">:</span> <span class="c1"># ratio ADref/DP == ADalt/DP == 0</span>
        <span class="c1"># do nothing: prob_homoRef = prob_homoAlt = prob_hetero = 0.0</span>
        <span class="k">pass</span>

    <span class="n">prob_geno</span> <span class="o">=</span> <span class="p">[</span><span class="n">prob_homoRef</span><span class="p">,</span> <span class="n">prob_hetero</span><span class="p">,</span> <span class="n">prob_homoAlt</span><span class="p">,</span> <span class="s1">&#39;NA&#39;</span><span class="p">]</span>
    <span class="k">if</span> <span class="n">depth</span> <span class="o">&gt;=</span> <span class="mi">1200</span><span class="o">*</span><span class="n">depth_division_th</span> <span class="p">:</span>  <span class="c1"># modif MY</span>
        <span class="c1"># Edit prob_geno</span>
        <span class="k">if</span> <span class="n">prob_homoRef</span> <span class="o">&gt;</span> <span class="n">prob_hetero</span> <span class="ow">and</span> <span class="n">prob_homoRef</span> <span class="o">&gt;</span> <span class="n">prob_homoAlt</span><span class="p">:</span>
            <span class="n">prob_geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">genoRef</span>
        <span class="k">elif</span> <span class="n">prob_homoAlt</span> <span class="o">&gt;</span> <span class="n">prob_hetero</span> <span class="ow">and</span> <span class="n">prob_homoAlt</span> <span class="o">&gt;</span> <span class="n">prob_homoRef</span><span class="p">:</span>
            <span class="n">prob_geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">genoAlt</span>
        <span class="k">elif</span> <span class="n">prob_hetero</span> <span class="o">&gt;</span> <span class="n">prob_homoRef</span> <span class="ow">and</span> <span class="n">prob_hetero</span> <span class="o">&gt;</span> <span class="n">prob_homoAlt</span><span class="p">:</span>
            <span class="n">prob_geno</span><span class="p">[</span><span class="mi">3</span><span class="p">]</span> <span class="o">=</span> <span class="n">genoRef</span> <span class="o">+</span> <span class="s2">&quot;/&quot;</span> <span class="o">+</span> <span class="n">genoAlt</span>
        <span class="k">else</span><span class="p">:</span> <span class="c1"># prob_homoRef == prob_homoAlt == heteroAB == 0.0</span>
            <span class="k">pass</span> <span class="c1"># do nothing: prob_geno = [0.0, 0.0, 0.0, &quot;NA&quot;]</span>
    <span class="k">else</span><span class="p">:</span>
        <span class="k">pass</span>
    <span class="n">cur_geno</span> <span class="o">=</span> <span class="p">[</span><span class="n">start</span><span class="p">,</span> <span class="n">stop</span><span class="p">,</span> <span class="n">ratio_ADref_DP</span><span class="p">,</span> <span class="n">ratio_ADalt_DP</span><span class="p">]</span> <span class="o">+</span> <span class="n">prob_geno</span>
    <span class="k">return</span> <span class="n">cur_geno</span>
</pre></div>
</div>
<p><strong>Explanation:</strong></p>
<p><cite>GetGenoWindow</cite>: This function determines the genotype of a genetic window based on the ratios of reads aligned to the reference and alternative alleles, as well as sequencing depth. It returns a list containing information about the genetic window, including probabilities of being homozygous for the reference and alternative alleles, the probability of being heterozygous, the predicted genotype, etc.
- <strong>Input:</strong></p>
<blockquote>
<div><ul class="simple">
<li><p><cite>cur_window</cite>: List containing elements of the window [start, stop, ADref, ADalt, DP].</p></li>
<li><p><cite>genoRef</cite>: Reference genotype.</p></li>
<li><p><cite>genoAlt</cite>: Alternative genotype.</p></li>
<li><p><cite>min_homo_freq</cite>: Minimum homozygosity frequency (defaulted to 0.75).</p></li>
</ul>
</div></blockquote>
<ul class="simple">
<li><p><strong>Functionality:</strong>
- The function first calculates the ADref/DP and ADalt/DP ratios.
- Then, it determines the probabilities of being homozygous for the reference genotype (<cite>prob_homoRef</cite>), homozygous for the alternative allele (<cite>prob_homoAlt</cite>), and heterozygous (<cite>prob_hetero</cite>) based on these ratios.
- Finally, it determines the genotype of the window based on the probabilities and returns it in <cite>cur_geno</cite>.</p></li>
<li><p><strong>Output:</strong>
- <cite>cur_geno</cite>: List containing the elements of the window along with the determined probabilities and genotype.</p></li>
</ul>
</section>
</section>

        </article>
      </div>
      <footer>
        
        <div class="related-pages">
          <a class="next-page" href="detectCOs_read_files.html">
              <div class="page-info">
                <div class="context">
                  <span>Next</span>
                </div>
                <div class="title">detectCOs_read_files.py</div>
              </div>
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
            </a>
          <a class="prev-page" href="config_detectCOs_CG.html">
              <svg class="furo-related-icon"><use href="#svg-arrow-right"></use></svg>
              <div class="page-info">
                <div class="context">
                  <span>Previous</span>
                </div>
                
                <div class="title">config_detectCOs_CG.yaml</div>
                
              </div>
            </a>
        </div>
        <div class="bottom-of-page">
          <div class="left-details">
            <div class="copyright">
                Copyright &#169; 2024, Mohamad Yassine
            </div>
            Made with <a href="https://www.sphinx-doc.org/">Sphinx</a> and <a class="muted-link" href="https://pradyunsg.me">@pradyunsg</a>'s
            
            <a href="https://github.com/pradyunsg/furo">Furo</a>
            
          </div>
          <div class="right-details">
            
          </div>
        </div>
        
      </footer>
    </div>
    <aside class="toc-drawer">
      
      
      <div class="toc-sticky toc-scroll">
        <div class="toc-title-container">
          <span class="toc-title">
            On this page
          </span>
        </div>
        <div class="toc-tree-container">
          <div class="toc-tree">
            <ul>
<li><a class="reference internal" href="#">detectCOs_required_functions.py</a><ul>
<li><a class="reference internal" href="#function-calculatehomoprobability">Function : CalculateHomoProbability</a></li>
<li><a class="reference internal" href="#function-calculateheteroprobability">Function : CalculateHeteroProbability</a></li>
<li><a class="reference internal" href="#function-getgenowindow">Function : GetGenoWindow</a></li>
<li><a class="reference internal" href="#function-updated-getgenowindow">Function : Updated GetGenoWindow</a></li>
</ul>
</li>
</ul>

          </div>
        </div>
      </div>
      
      
    </aside>
  </div>
</div><script data-url_root="./" id="documentation_options" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/_sphinx_javascript_frameworks_compat.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/sphinx_highlight.js"></script>
    <script src="_static/scripts/furo.js"></script>
    </body>
</html>