Search.setIndex({"docnames": ["Test_laucher_With_diff_SNPs", "Test_yaml_detectCOs_BC", "VisualiseGenotypesChromosome", "config_detectCOs_CG", "detectCOs_identifyCOs", "detectCOs_read_files", "detectCOs_required_functions", "detectCOs_sliding_window", "index", "test_detectCOs_polyrec_functions", "usage"], "filenames": ["Test_laucher_With_diff_SNPs.rst", "Test_yaml_detectCOs_BC.rst", "VisualiseGenotypesChromosome.rst", "config_detectCOs_CG.rst", "detectCOs_identifyCOs.rst", "detectCOs_read_files.rst", "detectCOs_required_functions.rst", "detectCOs_sliding_window.rst", "index.rst", "test_detectCOs_polyrec_functions.rst", "usage.rst"], "titles": ["Test_laucher_With_diff_SNPs.py", "Test_yaml_detectCOs_BC.yaml", "VisualiseGenotypesChromosome.py", "config_detectCOs_CG.yaml", "detectCOs_identifyCOs.py", "detectCOs_read_files.py", "detectCOs_required_functions.py", "detectCOs_sliding_window.py", "Welcome to detectCOs\u2019s documentation!", "test_detectCOs_polyrec_functions.py", "Usage"], "terms": {"thi": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10], "python": [0, 2, 3, 8, 9], "script": [0, 2, 9], "i": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10], "launcher": [0, 9], "detectco": [0, 3, 9, 10], "tool": [0, 3, 8, 9, 10], "It": [0, 2, 3, 4, 5, 6, 7, 9], "": [0, 1, 2, 4, 5, 7, 9, 10], "design": [0, 8, 9], "run": [0, 3, 9], "from": [0, 2, 3, 4, 5, 6, 7, 8, 9, 10], "command": [0, 9, 10], "line": [0, 2, 5, 9], "yaml": [0, 8, 9, 10], "configur": [0, 1, 9], "file": [0, 2, 3, 5, 8, 9, 10], "input": [0, 4, 5, 6, 7, 9], "import": [0, 4, 5, 6, 7, 9, 10], "necessari": [0, 3, 4, 5, 9, 10], "modul": [0, 9], "function": [0, 2, 8, 9], "packag": [0, 3, 9, 10], "includ": [0, 2, 6, 7, 8, 9, 10], "read": [0, 2, 5, 6, 7, 9], "perform": [0, 4, 9, 10], "slide": [0, 4, 7, 9], "window": [0, 1, 2, 4, 6, 7, 9], "analysi": [0, 1, 4, 5, 9, 10], "identifi": [0, 4, 5, 9, 10], "crossov": [0, 4, 8, 9], "event": [0, 4, 8, 9], "o": [0, 2, 5, 9, 10], "sy": [0, 5, 9], "pprint": [0, 9], "detectcos_read_fil": [0, 8, 9], "detectcos_sliding_window": [0, 8, 9], "detectcos_identifyco": [0, 8, 9], "provid": [0, 9], "usag": [0, 8, 9], "instruct": [0, 8, 9], "specifi": [0, 1, 2, 3, 4, 5, 7, 9, 10], "path": [0, 1, 2, 5, 9, 10], "an": [0, 2, 4, 5, 6, 7, 9], "argument": [0, 9], "load": [0, 5, 9], "creat": [0, 1, 2, 3, 4, 5, 9, 10], "output": [0, 2, 4, 5, 6, 7, 9, 10], "directori": [0, 1, 2, 5, 9], "The": [0, 1, 2, 4, 5, 6, 7, 9, 10], "begin": [0, 4, 5, 7], "format": [0, 5, 10], "If": [0, 4, 5, 6, 7], "rais": [0, 4, 5, 7, 9], "importerror": [0, 9], "current": [0, 2, 4, 7, 8, 9], "work": [0, 4, 9], "set": [0, 4, 5, 9], "root": [0, 2, 9], "polyrec": [0, 1, 9], "project": [0, 8, 9], "base": [0, 1, 2, 3, 4, 5, 6, 7], "A": [0, 4, 5, 7, 9], "variabl": [0, 4, 6], "outdir": [0, 2], "save": [0, 2, 4, 5, 6, 7], "all": [0, 1, 2, 3, 4, 5, 7, 10], "result": [0, 5, 6], "doesn": 0, "t": [0, 2, 4, 5, 7, 9], "exist": [0, 5], "print": [0, 2, 4, 5, 7, 9], "config": [0, 8, 9], "len": [0, 2, 4, 5, 9], "argv": [0, 9], "1": [0, 1, 2, 3, 4, 5, 6, 7, 9], "pleas": [0, 7, 9], "give": [0, 5, 9], "config_fil": [0, 9], "open": [0, 2, 5, 9], "r": [0, 2, 3, 5, 9, 10], "cf": [0, 9], "loader": [0, 9], "fullload": [0, 9], "chdir": [0, 9], "path_to_polyrec_project": [0, 1, 9], "output_dir_in_polyrec": [0, 1], "analyze_id": [0, 1, 5], "And": 0, "makedir": [0, 5], "chromosom": [0, 1, 2, 4, 5, 7, 8, 9], "length": [0, 1, 4, 5, 7], "ar": [0, 1, 2, 3, 4, 5, 6, 7, 9, 10], "n": [0, 2, 5, 7, 9], "step": [0, 1, 2, 4, 7, 8, 9], "chr_length": [0, 7], "readchrlen": [0, 8], "input_chr_len": [0, 5], "chr_len": [0, 1, 5, 7], "prefix_chr": [0, 1, 5, 9], "nchr_length": 0, "chr": [0, 1, 5, 7, 9], "lenght": 0, "list": [0, 2, 3, 4, 5, 6, 7, 9], "item": [0, 2, 4, 5, 7, 9], "posit": [0, 1, 4, 5, 7, 8, 9], "centromer": [0, 1, 4, 5, 7], "region": [0, 1, 4, 5, 7, 10], "each": [0, 2, 4, 5, 6, 7, 9, 10], "2": [0, 1, 2, 3, 4, 5, 7, 8, 9], "readcentroreg": [0, 8], "input_centro_reg": [0, 5], "centromere_reg": [0, 1], "ncentromer": 0, "left_bord": 0, "right_bord": 0, "parent": [0, 1, 2, 5, 7, 9, 10], "snp": [0, 1, 4, 5, 6, 7, 9, 10], "vcf": [0, 1, 5, 9, 10], "data": [0, 1, 2, 3, 4, 5, 7, 8, 9, 10], "variant": [0, 5, 8], "call": [0, 4, 5, 7, 8], "ha": [0, 4, 5, 7], "been": [0, 4, 5], "previous": [0, 4], "process": [0, 4, 5, 7], "otherwis": [0, 5, 7], "readparentalvcf": [0, 7, 8, 9], "last": [0, 4, 5, 7, 8], "per": [0, 4, 5, 7], "3": [0, 1, 3, 4, 5, 6, 7, 9], "parental_snp": [0, 5, 7, 9], "txt": [0, 1], "parental_snps_last_pos_chr": 0, "load_file_in_dict": [0, 8], "parental_last_snp_chr": [0, 9], "els": [0, 2, 4, 5, 6, 7], "input_vcf": [0, 5, 9], "parental_vcf": [0, 1, 9], "nparental_snp": [0, 9], "chr_po": [0, 5, 9], "po": [0, 4, 5, 7, 9], "gt": [0, 5, 7, 9], "5": [0, 2, 3, 4, 6, 7, 9], "nparental_last_snp_chr": [0, 9], "nsave": 0, "export_dict_in_fil": [0, 8], "my_dict": [0, 5], "output_fil": [0, 5], "header": [0, 5], "tchr": 0, "tpo": 0, "tgt": 0, "overwrit": [0, 5], "fals": [0, 4, 5], "offspr": [0, 1, 4, 5, 7], "readoffspringvcf": [0, 7, 8], "obtain": [0, 4], "4": [0, 3, 4, 7, 10], "offspring_snp": [0, 7], "offspring_snps_last_pos_chr": 0, "offspring_last_snp_chr": 0, "sample_vcf": [0, 1], "geno_ref": [0, 5, 7], "genotype_ref": [0, 1], "geno_alt": [0, 5, 7], "genotype_alt": [0, 1], "noffspring_snp": 0, "adref": [0, 4, 5, 6, 7], "adalt": [0, 4, 5, 6, 7], "genotyp": [0, 1, 2, 4, 5, 6, 7, 8, 9], "tadref": 0, "tadalt": 0, "tgenotyp": 0, "last_snps_chr": [0, 7], "tpos_last_snp": 0, "summari": [0, 4, 10], "info": [0, 4, 5, 7], "about": [0, 4, 5, 6, 7, 8, 9, 10], "parentalslidingwindow": [0, 8], "summar": 0, "parental_snps_window": [0, 7], "window_s": [0, 1, 4, 7], "nparental_snps_window": 0, "chr_window": [0, 2, 4, 7], "start": [0, 4, 5, 6, 7], "stop": [0, 2, 4, 6, 7], "nb_snp": 0, "2012": 0, "2017": 0, "tstart": 0, "tstop": 0, "tnb_snp": 0, "offspringslidingwindow": [0, 8], "6": [0, 3, 7], "offspring_snps_window": [0, 7], "noffspring_snps_window": 0, "dp": [0, 1, 4, 5, 6, 7], "nbsnp_aref": [0, 7], "nbsnp_aalt": [0, 7], "totsnp": [0, 7], "tdp": 0, "tnbsnp_aref": 0, "tnbsnp_aalt": 0, "ttot_snp": 0, "normal": [0, 7], "normalizeoffspringslidingwindow": [0, 4, 8], "relat": 0, "inform": [0, 1, 4, 5, 6, 7, 9, 10], "7": [0, 3, 4, 6, 7], "offspring_snps_window_norm": 0, "offspring_genotype_window_norm": 0, "offspring_nb_window_norm": 0, "nb_windows_chr": [0, 4, 7], "nb_window_chr": [0, 7], "min_snp_num": [0, 1, 7], "min_reads_num": [0, 1, 7], "ratio_min_homo": [0, 1, 7], "noffspring_snps_window_norm": 0, "noffspring_genotype_window_norm": 0, "probhomoref": [0, 4, 6], "probhetero": [0, 4, 6], "probhomoalt": [0, 4, 6], "width": 0, "120": 0, "compact": 0, "true": [0, 2, 4, 5], "nnb_windows_chr": 0, "tprobhomoref": 0, "tprobhetero": 0, "tprobhomoalt": 0, "tnb_window": 0, "8": [0, 3, 4, 10], "bi": 0, "visual": [0, 2], "exampl": 0, "section": [0, 8], "code": [0, 1, 3, 4, 5, 6, 7], "us": [0, 1, 2, 3, 4, 5, 6, 7, 9, 10], "map": [0, 2], "assign": [0, 2, 7], "color": [0, 2], "differ": [0, 4, 5, 6], "represent": [0, 2], "file_path": [0, 2], "directory_nam": [0, 2], "offspring_genotype_window_normalized_smooth": 0, "genotype_color": [0, 2], "blue": 0, "red": 0, "col": [0, 1], "ct": [0, 1], "green": 0, "na": [0, 2, 4, 5, 6, 7], "grai": [0, 2], "visualsize_genotop": [0, 8], "smooth": [0, 4, 7], "smoothnormalizedoffspringslidingwindow": [0, 4, 7], "separ": [0, 5], "offspring_snps_window_normalized_smooth": 0, "offspring_snps_window_smooth": 0, "offspring_genotype_window_smooth": [0, 4], "check_smooth": [0, 7], "smoothnormalizedosffspringslidingwindow": [0, 4, 8], "noffspring_snps_window_smooth": 0, "noffspring_genotype_window_normalized_smooth": 0, "ncheck_smooth": 0, "start_window": [0, 4, 7], "stop_window": [0, 7], "start_smooth": [0, 7], "stop_smooth": [0, 7], "window_smooth": 0, "tstart_window": 0, "tstop_window": 0, "tstart_smooth": 0, "tstop_smooth": 0, "twindow_smooth": 0, "co": [0, 1, 4, 10], "identifyco": [0, 8], "candid": [0, 4], "export": 0, "name": [0, 1, 2, 3, 5, 8], "candidateco": [0, 4], "qichao": [0, 8], "version": [0, 8], "anoth": [0, 4], "approach": 0, "refer": [0, 1, 4, 5, 6, 7, 8, 10], "implement": 0, "method": 0, "addit": [0, 4, 8], "probabl": [0, 4, 6, 7], "ratio": [0, 1, 4, 6, 7], "accord": [0, 4, 6, 7], "candidates_co_qichao": 0, "9": [0, 3, 4, 5, 6, 7], "candidates_co": [0, 4], "db_co": [0, 4], "ncandidateco": 0, "start_win": [0, 4], "stop_win": [0, 4], "co_start": [0, 4], "co_stop": [0, 4], "pre_geno": [0, 4], "cur_geno": [0, 4, 6], "chr_mean_win": 0, "tstart_win": 0, "tstart_co": 0, "tstop_co": 0, "tpre_geno": 0, "tcur_geno": 0, "offspring_smooth_prob": 0, "ordereddict": [0, 4, 5, 7], "kei": [0, 2, 4, 5, 7, 9], "valu": [0, 2, 4, 5, 7], "0": [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10], "offspring_sliding_geno_ratio": 0, "identifycosqichao": [0, 8], "offspring_smoothprob": [0, 4], "offspring_smoothwinnum": [0, 4], "offspring_slidinggenonum": [0, 4], "offspring_slidinggenoratio": [0, 4], "genoref": [0, 4, 6], "genoalt": [0, 4, 6], "10": [0, 1, 2, 4, 7], "pricis": 0, "new": [0, 1, 4, 5], "which": [0, 2, 4, 5, 8, 10], "adjust": [0, 4, 7], "newli": 0, "size": [0, 1, 2, 4, 5, 7], "precis": 0, "preciseco": [0, 8], "further": [0, 8], "record": [0, 4, 5, 7], "keep": 0, "db_co_new": 0, "rang": [0, 2, 4, 5, 7], "start_co": 0, "100000": 0, "end_co": 0, "split": [0, 2, 4, 5, 7], "_": [0, 2, 4, 5, 7], "new_offspring_genotype_window_smooth": 0, "cle": 0, "valeur": 0, "offspring_genotype_window_smoothed_new": 0, "int": [0, 4, 5, 7], "end_window": [0, 4], "candidates_co_2": 0, "db_co_2": 0, "cur_chr": [0, 4, 7], "str": [0, 2, 4, 5, 6, 7], "round": [0, 4, 6, 7], "updat": [0, 4, 5, 7, 8], "updated_preciseco": 0, "modifi": [0, 1, 4, 7], "my": [0, 1, 4, 6, 7], "detail": [0, 4, 5], "start_win_stop_win": 0, "diff": [0, 4], "calcul": [0, 1, 4, 6, 7], "de": [0, 8], "la": 0, "diff\u00e9renc": 0, "precisecos_": 0, "new_window_s": [0, 1], "_kb": 0, "stop_co": 0, "refin": [0, 1, 4], "border": [0, 4, 5, 7], "also": [0, 4, 5, 7], "refinecobordersqichao": [0, 8], "refinedco": [0, 4], "nrefinecobord": 0, "offspring_refinedco": 0, "refinecobord": [0, 8], "offspring_refinedcos_qichao": 0, "accompani": 0, "explanatori": 0, "comment": [0, 5], "statement": 0, "track": [0, 4], "progress": 0, "variou": [1, 7, 9], "requir": [1, 3, 4, 7, 9], "well": [1, 4, 5, 6], "paramet": [1, 2, 4, 7, 9], "mandatori": 1, "gener": [1, 2, 10], "folder": [1, 10], "sampl": [1, 10], "your": 1, "parental_col": 1, "ler_offspring_plate1_a11": 1, "prefix": [1, 5], "ad": [1, 5, 6, 7], "chrom": 1, "column": [1, 2, 5], "offspring_vcf": 1, "thei": [1, 4, 6, 7], "have": [1, 4, 7, 8], "same": [1, 4], "between": [1, 4, 5, 7], "two": [1, 4, 5, 6], "chr_len_athaliana_qichao": 1, "centromere_athaliana_qichao": 1, "parental_golden_snps_139_genotyped_indiv_28_merged_formatted_idpoint_nochr_dict_format": 1, "111355_aa_gt_by_golden_awk": 1, "block": 1, "altern": [1, 4, 5, 6, 7], "kb": [1, 4, 7], "automat": [1, 7], "calculta": 1, "half": [1, 7], "default": [1, 2, 3, 4, 5, 6, 7], "100": [1, 2, 7], "minimum": [1, 4, 6, 7], "number": [1, 2, 4, 5, 6, 7], "coverag": [1, 5, 7], "determin": [1, 4, 5, 6, 7], "16": [1, 7], "frequenc": [1, 4, 6, 7], "homozyg": [1, 4, 5, 6, 7], "float": [1, 4, 5, 6, 7], "exclud": [1, 7], "75": [1, 2, 6, 7], "depth": [1, 5, 6, 7], "divis": 1, "when": [1, 4, 7], "depth_division_th": [1, 6], "To": [1, 10], "new_depth_division_th": 1, "allow": [1, 2, 4, 5, 7, 10], "custom": 1, "specif": [1, 2, 4, 5, 6, 9, 10], "analyz": [1, 4, 7, 8, 10], "sever": [2, 4], "util": 2, "handl": [2, 4], "genom": [2, 10], "support": [2, 4, 5, 6], "oper": [2, 4], "like": 2, "delimit": [2, 5], "detect": [2, 4, 8, 10], "csv": 2, "search": [2, 3, 4], "def": [2, 4, 5, 6, 7], "num_lin": 2, "given": [2, 6], "first": [2, 4, 5, 6, 7, 9, 10], "check": [2, 4, 5, 7, 8], "common": 2, "comma": 2, "semicolon": 2, "tab": [2, 5], "space": [2, 5], "param": [2, 4], "return": [2, 4, 5, 6, 7], "newlin": [2, 5], "readlin": 2, "count": [2, 4, 9], "occurr": 2, "potenti": [2, 4, 5], "delimiter_count": 2, "delim": 2, "counter": [2, 5, 7], "highest": 2, "detected_delimit": 2, "max": [2, 7], "get": [2, 5], "examin": [2, 4], "few": [2, 9], "pipe": 2, "dictionari": [2, 4, 5, 7], "nan": [2, 6], "convert": [2, 5, 6, 7], "them": [2, 4, 5], "datafram": 2, "contain": [2, 4, 5, 6, 7, 9, 10], "where": [2, 3, 4, 5, 6, 10], "correspond": [2, 4, 5, 7], "copi": 2, "string": [2, 4, 5], "explicitli": 2, "appli": 2, "lambda": 2, "x": [2, 4, 5], "pd": 2, "isnul": 2, "strip": [2, 5], "ensur": [2, 3, 7, 10], "appropri": [2, 5], "filenam": 2, "max_dir": 2, "none": [2, 4], "through": [2, 4, 5, 7], "subdirectori": 2, "wa": 2, "found": [2, 4, 5], "full": 2, "limit": [2, 10], "consid": [2, 4, 5, 7, 10], "maximum": [2, 7], "unlimit": 2, "dir_count": 2, "travers": 2, "subdir": 2, "dir": 2, "walk": 2, "skip": [2, 5], "continu": [2, 4, 5], "desir": 2, "add": [2, 4, 5], "basenam": 2, "join": [2, 5], "reach": 2, "break": [2, 4], "particular": 2, "chromosome_data_all_fil": 2, "f": [2, 4, 10], "read_csv": 2, "sep": [2, 4], "head": 2, "expand": 2, "uniqu": [2, 5, 7], "chromosome_data": 2, "startswith": [2, 5], "chromosome_data_with_color": 2, "reset_index": 2, "drop": 2, "append": [2, 5], "data_list": 2, "50": 2, "plt": 2, "figur": 2, "figsiz": 2, "20": 2, "fontsiz": 2, "12": 2, "elif": [2, 4, 5, 6, 7], "80": 2, "15": 2, "14": [2, 3], "60": 2, "30": 2, "18": 2, "y_posit": 2, "y": 2, "enumer": 2, "y_po": 2, "scatter": 2, "c": 2, "marker": [2, 5, 9], "500": 2, "label": 2, "xlabel": 2, "index": 2, "xtick": 2, "iloc": 2, "rotat": 2, "90": 2, "ylim": 2, "ytick": 2, "tight_layout": 2, "subplots_adjust": 2, "right": [2, 5, 7], "dynam": 2, "legend": 2, "element": [2, 5, 6, 7], "legend_el": 2, "line2d": 2, "w": [2, 5], "markerfacecolor": 2, "markers": 2, "loc": 2, "center": [2, 4], "left": [2, 5], "bbox_to_anchor": 2, "titl": 2, "savefig": 2, "png": 2, "dpi": 2, "bbox_inch": 2, "tight": 2, "show": 2, "organ": 2, "store": [2, 4, 5, 7], "matplotlib": 2, "plot": 2, "imag": 2, "conda": [3, 10], "environ": [3, 10], "channel": 3, "forg": 3, "bioconda": 3, "atgc": 3, "montpelli": 3, "depend": 3, "pip": 3, "23": 3, "hue": [3, 4, 5, 7], "scipi": [3, 6], "11": 3, "pyyaml": 3, "chang": [3, 4], "resolv": 3, "conflict": 3, "dplyr": 3, "stringr": 3, "tabl": 3, "unix": 3, "defin": [3, 4, 5, 6, 7], "instal": [3, 8], "consist": [3, 7], "reproduc": 3, "manner": 3, "collect": [4, 5, 7], "detectcos_required_funct": [4, 5, 7, 8], "descript": [4, 5, 6, 7], "find": 4, "locat": 4, "hypothet": 4, "cross": [4, 10], "over": [4, 7, 10], "order": [4, 5, 7], "dictionnari": [4, 5, 7], "chr_mean_window": 4, "left_window_bord": 4, "right_window_bord": 4, "previou": [4, 7], "doubl": 4, "dure": 4, "even": [4, 7], "under": [4, 8], "threshold": 4, "thr_support": 4, "start_co_win": 4, "stop_co_win": 4, "statu": 4, "upgrad": 4, "300kb": 4, "nb_win": [4, 7], "thr_support_kb": 4, "homo": [4, 5], "enough": 4, "iter": [4, 5, 7], "num_win": [4, 7], "cur_support": 4, "geno_stop": 4, "cur_win": 4, "geonotyp": 4, "befor": [4, 10], "pre_win": 4, "count_na": 4, "consecut": 4, "support_prena": 4, "geno_prena": 4, "initi": [4, 5, 7, 8, 9], "penultim": 4, "due": [4, 5], "key_start": 4, "geno_start": 4, "key_stop": 4, "edit": [4, 6], "min": 4, "valid": [4, 5, 7], "other": [4, 5], "posist": 4, "mean": [4, 5], "end": [4, 5, 7], "enter": [4, 7], "win_prena": 4, "voir": 4, "ca": 4, "13": 4, "modif": [4, 6], "reset": 4, "least": [4, 7], "pre_num": 4, "useless": 4, "key_db_co": 4, "not_support": 4, "some": 4, "warn": [4, 5, 7], "Be": 4, "care": 4, "explan": [4, 5, 6, 7], "produc": 4, "deriv": 4, "suffici": 4, "averag": 4, "after": [4, 5], "do": [4, 6, 7, 10], "need": [4, 5], "inclus": 4, "setup": 4, "within": [4, 7], "relative_posit": 4, "relative_end_posit": 4, "scenario": 4, "logic": 4, "genet": [4, 6, 8, 9, 10], "coordin": [4, 5], "predefin": 4, "offspring_infosnp": 4, "co_kei": 4, "co_valu": 4, "cur_po": 4, "win": [4, 7], "cur_start": 4, "cur_stop": 4, "1000": [4, 7], "cur_geno1": 4, "cur_geno2": 4, "pre_geno_3": 4, "cur_geno_3": 4, "comm_geno": 4, "refined_infosnp": 4, "refined_infosnp_cnt": 4, "refined_infosnp_po": 4, "snp_geno": 4, "refined_info_start": 4, "refined_info_stop": 4, "snp_po": 4, "snp_type": 4, "2000": 4, "error": [4, 5, 7], "refinedinfoco": 4, "pre_geno1": 4, "pre_geno2": 4, "refined_totalsnp": 4, "refined_totalsnp_cnt": 4, "re": [4, 5], "re_refinedco": 4, "pre_co_chr": 4, "pre_co_po": 4, "pre_co_start": 4, "pre_co_stop": 4, "pre_co_geno1": 4, "pre_co_geno2": 4, "cur_item": 4, "co_chr": 4, "co_po": 4, "co_geno1": 4, "co_geno2": 4, "pass": [4, 5, 6], "geno": [4, 5], "were": [4, 8], "take": [4, 7], "three": [4, 5, 7], "entri": [4, 5, 7, 9], "pair": [4, 7, 10], "repres": [4, 5], "singl": [4, 10], "nucleotid": [4, 10], "polymorph": [4, 10], "empti": [4, 5], "next": [4, 7], "follow": [4, 5], "extract": [4, 5, 7, 9], "For": [4, 5, 7, 8], "match": [4, 5, 9], "win_num": 4, "min_win_num": 4, "max_win_num": 4, "cur_num": 4, "pre_po": 4, "cur1_geno": 4, "cur1_pos1": 4, "cur1_pos2": 4, "cur1_num": 4, "key1": 4, "geno1": 4, "key2": 4, "geno2": 4, "should": 4, "happen": 4, "onli": [4, 5, 10], "final": [4, 5, 6, 7], "close": 4, "300000": 4, "larg": 4, "breakpoint": 4, "interv": 4, "correct": [4, 5, 7, 8], "cor_pre_po": 4, "cor_cur_po": 4, "rawco": 4, "correctco": 4, "point": [4, 5], "indic": [4, 5], "condit": 4, "paus": 4, "too": 4, "appear": [4, 10], "mai": 4, "recombin": [4, 5], "min_homo_freq": [4, 6], "correctedco": 4, "rawco_start": 4, "rawco_stop": 4, "rawco_geno1": 4, "rawco_geno2": 4, "cur_chr_centro_left": 4, "150": 4, "cur_chr_centro_right": 4, "pre_na_geno": 4, "pre_na_po": 4, "pre_na_num": 4, "getgenowindow": [4, 7, 8], "keepright": 4, "needonemor": 4, "yet": 4, "here": [4, 10], "its": [4, 5], "raw": [4, 5], "homozygot": 4, "retriev": [4, 5, 7], "adjac": 4, "ident": [4, 5], "case": [4, 10], "heterozyg": [4, 5, 6, 7], "both": [4, 5, 6], "move": [4, 10], "merg": 4, "overlap": [4, 7], "veri": 4, "correctli": 4, "input_fil": 5, "log": [5, 7], "getsiz": 5, "exit": 5, "Not": 5, "greater": [5, 7], "than": [5, 7], "zero": 5, "displai": 5, "messag": 5, "program": 5, "doe": 5, "No": 5, "explicit": 5, "consol": 5, "verif": [5, 7], "termin": 5, "old": 5, "checkouput": 5, "remov": 5, "delet": [5, 10], "Then": [5, 6], "value_col0": 5, "value_col1": 5, "value_coln": 5, "values_list": 5, "value_list": 5, "regex_integ": 5, "regex_float": 5, "regex_int": 5, "d": [5, 9], "equal": [5, 7], "fullmatch": 5, "filter": 5, "syntax": 5, "integ": [5, 7], "pattern": 5, "respect": 5, "one": [5, 7], "being": [5, 6], "entir": 5, "onc": 5, "either": 5, "directli": 5, "present": [5, 7, 10], "bool": 5, "must": [5, 7], "tabul": 5, "alreadi": 5, "write": 5, "isinst": 5, "boolean": 5, "whether": [5, 7], "By": 5, "overwritten": 5, "option": [5, 10], "enabl": [5, 10], "mode": [5, 10], "written": 5, "text": 5, "prefix_str": 5, "e": 5, "g": 5, "etc": [5, 6], "chr_x": 5, "valueerror": [5, 7], "invalid": [5, 7], "concaten": 5, "so": [5, 7], "without": 5, "applic": 5, "coordon": 5, "pericentromer": 5, "ignor": 5, "last_snp": 5, "combin": 5, "ninth": 5, "form": 5, "info_snp": 5, "info_last_snp": 5, "prepar": 5, "log_path": 5, "dirnam": 5, "__file__": 5, "readoffsrpingvcf_new_snp": 5, "readoffsrpingvcf_weird_snp": 5, "total_snp_num": 5, "weird_snp": 5, "new_snp": 5, "count_dp0": 5, "field": 5, "ref": [5, 6], "alt": [5, 6, 7], "ref_supp": 5, "alt_supp": 5, "try": 5, "except": 5, "fileexistserror": 5, "logfil": 5, "kept": 5, "weird": 5, "compos": 5, "allel": [5, 6, 7, 10], "among": 5, "second": 5, "id": 5, "total": [5, 7], "strang": 5, "sequenc": [5, 6, 8], "conform": 5, "increment": [5, 7], "readoffspringvcf_weird_snp": 5, "readoffspringvcf_new_snp": 5, "retain": 5, "snps_list": [5, 9], "snps_list2": 5, "snp_kei": 5, "snp_key2": 5, "em": [5, 9], "ethyl": 5, "methan": 5, "sulfon": 5, "piec": 5, "alt_list": 5, "snps_list_line1": [5, 9], "snps_list_line2": [5, 9], "geno_line1": 5, "geno_line2": 5, "final_geno": 5, "gt_list": 5, "associ": [5, 7], "miss": 5, "signifi": 5, "cannot": 5, "ani": 5, "simpli": 5, "input_ems_line1": 5, "input_ems_line2": 5, "unknown_snp": 5, "unknown": 5, "journal": 5, "descend": 5, "imput": 5, "math": 6, "numpi": 6, "np": 6, "decim": 6, "b": 6, "ada": 6, "adb": 6, "prob": 6, "special": 6, "comb": 6, "pow": 6, "99": 6, "01": 6, "isnan": 6, "isinf": 6, "nan_to_num": 6, "binomi": 6, "law": 6, "formula": 6, "divid": [6, 7], "coeffici": 6, "choos": 6, "multipli": 6, "success": 6, "failur": 6, "non": 6, "finit": 6, "inf": [6, 7], "estim": 6, "homozygos": 6, "heterozygos": 6, "cur_window": [6, 7], "min_hetero_freq": 6, "ratio_adref_dp": 6, "ratio_adalt_dp": [6, 7], "prob_homoref": 6, "prob_homoalt": 6, "prob_hetero": [6, 7], "noth": 6, "prob_geno": 6, "heteroab": 6, "align": 6, "predict": 6, "clear": 6, "along": [6, 7], "depth_divis": 6, "1200": 6, "100kb": 7, "1mb": 7, "1000kb": 7, "snps_window": 7, "chr_num": 7, "nb": [7, 9], "avoid": 7, "calcult": 7, "pb": 7, "twindow": 7, "last_snp_po": 7, "note": 7, "total_snp": 7, "kilobas": 7, "mb": 7, "verifi": 7, "less": 7, "bp": 7, "eas": 7, "up": 7, "nearest": 7, "exce": 7, "sum": 7, "paramat": 7, "studi": [7, 8], "nbsnpref": 7, "nbsnpalt": 7, "homoa": 7, "heteab": 7, "homob": 7, "similar": 7, "instead": 7, "statist": 7, "dertermin": 7, "main": 7, "geno_window": 7, "ratio_adref": 7, "prob_homo_ref": 7, "prob_homo_alt": 7, "probhomoa": 7, "probheteroab": 7, "probhomob": 7, "nb_window": 7, "key_window": 7, "ad_ref": 7, "ad_alt": 7, "NOT": 7, "nb_snps_window": 7, "num_last_window": 7, "win_id": 7, "predomin": 7, "exclus": 7, "becaus": 7, "dictionnai": 7, "account": 7, "insid": 7, "But": 7, "we": 7, "ratio_min_hetero": 7, "nb_win_chr": 7, "cen_left": 7, "cen_right": 7, "start_win_smooth": 7, "stop_win_smooth": 7, "pos_start_cur_win": 7, "pos_stop_cur_win": 7, "offspring_slidinggenowindow": 7, "pos_stop_cur_wind": 7, "neighbor": 7, "boundari": 7, "minu": 7, "plu": 7, "bound": 7, "intersect": 7, "more": 8, "wikipedia": 8, "activ": [8, 10], "develop": 8, "out": 8, "how": 8, "author": 8, "lian": 8, "email": 8, "qlian": 8, "mpipz": 8, "mpg": 8, "corrector": 8, "ma\u00ebla": 8, "s\u00e9meri": 8, "maela": 8, "semeri": 8, "inra": 8, "fr": 8, "mohamad": 8, "yassin": 8, "releas": 8, "date": 8, "06": 8, "juli": 8, "2022": 8, "22": 8, "april": 8, "2024": 8, "cover": 8, "aim": 8, "enhanc": 8, "our": 8, "abil": 8, "interpret": 8, "These": [8, 10], "contribut": 8, "team": 8, "dedic": 8, "research": 8, "programm": 8, "who": 8, "carefulli": 8, "improv": [8, 10], "upon": 8, "launch": 8, "preprocess": 8, "gatk": 8, "haplotypecal": 8, "config_detectcos_cg": [8, 10], "py": [8, 10], "calculatehomoprob": 8, "calculateheteroprob": 8, "checkinput": 8, "removefil": 8, "reademslin": [8, 9], "identifygeno": 8, "readrecombinedoffspringvcf": 8, "correctlargecosqichao": 8, "test_detectcos_polyrec_funct": 8, "test_laucher_with_diff_snp": [8, 10], "test_yaml_detectcos_bc": 8, "FOR": 8, "visualisegenotypeschromosom": 8, "detect_delimit": 8, "assign_window_color": 8, "find_files_in_directori": 8, "pos_snps_a": 9, "ems_line1": 9, "pos_snps_d": 9, "ems_line2": 9, "nems_line1": 9, "nems_line2": 9, "count_a": 9, "count_d": 9, "golden": 9, "overal": 9, "serv": [9, 10], "central": 9, "compon": 9, "pipelin": 9, "feedback": 9, "virtual": 10, "mamba": 10, "env": 10, "p": 10, "miniconda3": 10, "detectcos_env": 10, "python3": 10, "crucial": 10, "individu": 10, "popul": 10, "toolkit": 10, "why": 10, "identif": 10, "indel": 10, "insert": 10, "dna": 10, "compar": 10, "help": 10, "sensit": 10, "model": 10, "known": 10, "confirm": 10, "interest": 10, "select": 10, "l": 10, "In": 10, "reduc": 10, "comput": 10, "burden": 10, "relev": 10, "emit_all_active_sit": 10, "site": 10, "emit": 10, "those": 10, "vari": 10, "comprehens": 10, "accur": 10, "basi": 10, "subsequ": 10, "java": 10, "jar": 10, "local": 10, "nativ": 10, "hmm": 10, "thread": 10, "tair10_chr_al": 10, "fasta": 10, "7rv498_recalibrated_sort": 10, "bam": 10, "7rv498_golden_snp": 10, "parental_goldensnps_indiv_28": 10}, "objects": {}, "objtypes": {}, "objnames": {}, "titleterms": {"test_laucher_with_diff_snp": 0, "py": [0, 2, 4, 5, 6, 7, 9], "test_yaml_detectcos_bc": 1, "yaml": [1, 3], "config": 1, "file": 1, "FOR": 1, "detectco": [1, 8], "tool": 1, "visualisegenotypeschromosom": 2, "detect_delimit": 2, "assign_window_color": 2, "find_files_in_directori": 2, "visualsize_genotop": 2, "config_detectcos_cg": 3, "detectcos_identifyco": 4, "function": [4, 5, 6, 7], "identifyco": 4, "preciseco": 4, "refinecobord": 4, "identifycosqichao": 4, "correctlargecosqichao": 4, "refinecobordersqichao": 4, "detectcos_read_fil": 5, "checkinput": 5, "removefil": 5, "load_file_in_dict": 5, "export_dict_in_fil": 5, "readchrlen": 5, "readcentroreg": 5, "readparentalvcf": 5, "readoffspringvcf": 5, "reademslin": 5, "identifygeno": 5, "readrecombinedoffspringvcf": 5, "detectcos_required_funct": 6, "calculatehomoprob": 6, "calculateheteroprob": 6, "getgenowindow": 6, "updat": 6, "detectcos_sliding_window": 7, "parentalslidingwindow": 7, "offspringslidingwindow": 7, "normalizeoffspringslidingwindow": 7, "smoothnormalizedosffspringslidingwindow": 7, "welcom": 8, "": 8, "document": 8, "program": 8, "inform": 8, "overview": 8, "content": 8, "test_detectcos_polyrec_funct": 9, "usag": 10, "instal": 10, "launch": 10, "preprocess": 10, "step": 10, "variant": 10, "call": 10, "gatk": 10, "haplotypecal": 10}, "envversion": {"sphinx.domains.c": 2, "sphinx.domains.changeset": 1, "sphinx.domains.citation": 1, "sphinx.domains.cpp": 8, "sphinx.domains.index": 1, "sphinx.domains.javascript": 2, "sphinx.domains.math": 2, "sphinx.domains.python": 3, "sphinx.domains.rst": 2, "sphinx.domains.std": 2, "sphinx": 57}, "alltitles": {"Test_laucher_With_diff_SNPs.py": [[0, "test-laucher-with-diff-snps-py"]], "Test_yaml_detectCOs_BC.yaml": [[1, "test-yaml-detectcos-bc-yaml"]], "CONFIG FILE FOR detectCOs TOOLS": [[1, "config-file-for-detectcos-tools"]], "VisualiseGenotypesChromosome.py": [[2, "visualisegenotypeschromosome-py"]], "detect_delimiter": [[2, "detect-delimiter"]], "assign_window_colors": [[2, "assign-window-colors"]], "find_files_in_directories": [[2, "find-files-in-directories"]], "visualsize_genotope": [[2, "visualsize-genotope"]], "config_detectCOs_CG.yaml": [[3, "config-detectcos-cg-yaml"]], "detectCOs_identifyCOs.py": [[4, "detectcos-identifycos-py"]], "Function : IdentifyCOs": [[4, "function-identifycos"]], "Function : PreciseCOs": [[4, "function-precisecos"]], "Function : RefineCOBorders": [[4, "function-refinecoborders"]], "Function : IdentifyCOsQichao": [[4, "function-identifycosqichao"]], "Function : CorrectLargeCOsQichao": [[4, "function-correctlargecosqichao"]], "Function : RefineCOBordersQichao": [[4, "function-refinecobordersqichao"]], "detectCOs_read_files.py": [[5, "detectcos-read-files-py"]], "Function : CheckInput": [[5, "function-checkinput"]], "Function : RemoveFile": [[5, "function-removefile"]], "Function : load_file_in_dict": [[5, "function-load-file-in-dict"]], "Function : export_dict_in_file": [[5, "function-export-dict-in-file"]], "Function : ReadChrLen": [[5, "function-readchrlen"]], "Function : ReadCentroReg": [[5, "function-readcentroreg"]], "Function : ReadParentalVCF": [[5, "function-readparentalvcf"]], "Function : ReadOffspringVCF": [[5, "function-readoffspringvcf"]], "Function : ReadEMSline": [[5, "function-reademsline"]], "Function : IdentifyGeno": [[5, "function-identifygeno"]], "Function : ReadRecombinedOffspringVCF": [[5, "function-readrecombinedoffspringvcf"]], "detectCOs_required_functions.py": [[6, "detectcos-required-functions-py"]], "Function : CalculateHomoProbability": [[6, "function-calculatehomoprobability"]], "Function : CalculateHeteroProbability": [[6, "function-calculateheteroprobability"]], "Function : GetGenoWindow": [[6, "function-getgenowindow"]], "Function : Updated GetGenoWindow": [[6, "function-updated-getgenowindow"]], "detectCOs_sliding_window.py": [[7, "detectcos-sliding-window-py"]], "Function : ParentalSlidingWindow": [[7, "function-parentalslidingwindow"]], "Function : OffspringSlidingWindow": [[7, "function-offspringslidingwindow"]], "Function : NormalizeOffspringSlidingWindow": [[7, "function-normalizeoffspringslidingwindow"]], "Function : SmoothNormalizedOsffspringSlidingWindow": [[7, "function-smoothnormalizedosffspringslidingwindow"]], "Welcome to detectCOs\u2019s documentation!": [[8, "welcome-to-detectcos-s-documentation"]], "Program Information": [[8, "program-information"]], "Overview": [[8, "overview"]], "Contents": [[8, "contents"]], "test_detectCOs_polyrec_functions.py": [[9, "test-detectcos-polyrec-functions-py"]], "Usage": [[10, "usage"]], "Installation": [[10, "installation"]], "launch": [[10, "launch"]], "Preprocessing Step: Variant Calling with GATK HaplotypeCaller": [[10, "preprocessing-step-variant-calling-with-gatk-haplotypecaller"]]}, "indexentries": {}})